{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "q1GUmQYBlKwz",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "q1GUmQYBlKwz",
    "outputId": "817c316a-01e2-4bdf-f51f-6ee5e7be8b8f"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: glycowork in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (1.2.0)\n",
      "Requirement already satisfied: scikit-learn in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork) (1.4.2)\n",
      "Requirement already satisfied: regex in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork) (2024.4.16)\n",
      "Requirement already satisfied: networkx>=3.0 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork) (3.2.1)\n",
      "Requirement already satisfied: statsmodels in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork) (0.14.2)\n",
      "Requirement already satisfied: scipy in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork) (1.10.1)\n",
      "Requirement already satisfied: seaborn in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork) (0.13.2)\n",
      "Requirement already satisfied: xgboost in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork) (2.0.3)\n",
      "Requirement already satisfied: mpld3 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork) (0.5.10)\n",
      "Requirement already satisfied: pandas>=1.3 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork) (1.5.3)\n",
      "Requirement already satisfied: matplotlib-inline in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork) (0.1.7)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from pandas>=1.3->glycowork) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from pandas>=1.3->glycowork) (2024.1)\n",
      "Requirement already satisfied: numpy>=1.20.3 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from pandas>=1.3->glycowork) (1.26.4)\n",
      "Requirement already satisfied: traitlets in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from matplotlib-inline->glycowork) (5.14.2)\n",
      "Requirement already satisfied: jinja2 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from mpld3->glycowork) (3.1.3)\n",
      "Requirement already satisfied: matplotlib in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from mpld3->glycowork) (3.7.5)\n",
      "Requirement already satisfied: joblib>=1.2.0 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from scikit-learn->glycowork) (1.4.0)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from scikit-learn->glycowork) (3.4.0)\n",
      "Requirement already satisfied: patsy>=0.5.6 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from statsmodels->glycowork) (0.5.6)\n",
      "Requirement already satisfied: packaging>=21.3 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from statsmodels->glycowork) (24.0)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from matplotlib->mpld3->glycowork) (1.2.1)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from matplotlib->mpld3->glycowork) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from matplotlib->mpld3->glycowork) (4.51.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from matplotlib->mpld3->glycowork) (1.4.5)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from matplotlib->mpld3->glycowork) (10.3.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from matplotlib->mpld3->glycowork) (3.1.2)\n",
      "Requirement already satisfied: importlib-resources>=3.2.0 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from matplotlib->mpld3->glycowork) (6.4.0)\n",
      "Requirement already satisfied: six in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from patsy>=0.5.6->statsmodels->glycowork) (1.16.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from jinja2->mpld3->glycowork) (2.1.5)\n",
      "Requirement already satisfied: zipp>=3.1.0 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from importlib-resources>=3.2.0->matplotlib->mpld3->glycowork) (3.18.1)\n",
      "Requirement already satisfied: glycowork[ml] in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (1.2.0)\n",
      "Requirement already satisfied: scikit-learn in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork[ml]) (1.4.2)\n",
      "Requirement already satisfied: regex in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork[ml]) (2024.4.16)\n",
      "Requirement already satisfied: networkx>=3.0 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork[ml]) (3.2.1)\n",
      "Requirement already satisfied: statsmodels in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork[ml]) (0.14.2)\n",
      "Requirement already satisfied: scipy in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork[ml]) (1.10.1)\n",
      "Requirement already satisfied: seaborn in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork[ml]) (0.13.2)\n",
      "Requirement already satisfied: xgboost in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork[ml]) (2.0.3)\n",
      "Requirement already satisfied: mpld3 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork[ml]) (0.5.10)\n",
      "Requirement already satisfied: pandas>=1.3 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork[ml]) (1.5.3)\n",
      "Requirement already satisfied: matplotlib-inline in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork[ml]) (0.1.7)\n",
      "Requirement already satisfied: torch-geometric in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork[ml]) (2.4.0)\n",
      "Requirement already satisfied: torch in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from glycowork[ml]) (2.1.0+cu118)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from pandas>=1.3->glycowork[ml]) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from pandas>=1.3->glycowork[ml]) (2024.1)\n",
      "Requirement already satisfied: numpy>=1.20.3 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from pandas>=1.3->glycowork[ml]) (1.26.4)\n",
      "Requirement already satisfied: traitlets in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from matplotlib-inline->glycowork[ml]) (5.14.2)\n",
      "Requirement already satisfied: jinja2 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from mpld3->glycowork[ml]) (3.1.3)\n",
      "Requirement already satisfied: matplotlib in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from mpld3->glycowork[ml]) (3.7.5)\n",
      "Requirement already satisfied: joblib>=1.2.0 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from scikit-learn->glycowork[ml]) (1.4.0)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from scikit-learn->glycowork[ml]) (3.4.0)\n",
      "Requirement already satisfied: patsy>=0.5.6 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from statsmodels->glycowork[ml]) (0.5.6)\n",
      "Requirement already satisfied: packaging>=21.3 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from statsmodels->glycowork[ml]) (24.0)\n",
      "Requirement already satisfied: filelock in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from torch->glycowork[ml]) (3.13.4)\n",
      "Requirement already satisfied: typing-extensions in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from torch->glycowork[ml]) (4.11.0)\n",
      "Requirement already satisfied: sympy in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from torch->glycowork[ml]) (1.12)\n",
      "Requirement already satisfied: fsspec in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from torch->glycowork[ml]) (2024.3.1)\n",
      "Requirement already satisfied: triton==2.1.0 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from torch->glycowork[ml]) (2.1.0)\n",
      "Requirement already satisfied: tqdm in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from torch-geometric->glycowork[ml]) (4.66.2)\n",
      "Requirement already satisfied: requests in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from torch-geometric->glycowork[ml]) (2.31.0)\n",
      "Requirement already satisfied: pyparsing in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from torch-geometric->glycowork[ml]) (3.1.2)\n",
      "Requirement already satisfied: psutil>=5.8.0 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from torch-geometric->glycowork[ml]) (5.9.8)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from matplotlib->mpld3->glycowork[ml]) (1.2.1)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from matplotlib->mpld3->glycowork[ml]) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from matplotlib->mpld3->glycowork[ml]) (4.51.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from matplotlib->mpld3->glycowork[ml]) (1.4.5)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from matplotlib->mpld3->glycowork[ml]) (10.3.0)\n",
      "Requirement already satisfied: importlib-resources>=3.2.0 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from matplotlib->mpld3->glycowork[ml]) (6.4.0)\n",
      "Requirement already satisfied: six in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from patsy>=0.5.6->statsmodels->glycowork[ml]) (1.16.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from jinja2->mpld3->glycowork[ml]) (2.1.5)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from requests->torch-geometric->glycowork[ml]) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from requests->torch-geometric->glycowork[ml]) (3.7)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from requests->torch-geometric->glycowork[ml]) (2.2.1)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from requests->torch-geometric->glycowork[ml]) (2024.2.2)\n",
      "Requirement already satisfied: mpmath>=0.19 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from sympy->torch->glycowork[ml]) (1.3.0)\n",
      "Requirement already satisfied: zipp>=3.1.0 in /home/jshee/miniconda3/envs/esm/lib/python3.9/site-packages (from importlib-resources>=3.2.0->matplotlib->mpld3->glycowork[ml]) (3.18.1)\n"
     ]
    }
   ],
   "source": [
    "# from google.colab import drive\n",
    "# drive.mount('/content/drive')\n",
    "fs = '/content/drive/My Drive/'\n",
    "fs = 'data/'\n",
    "\n",
    "!pip install glycowork\n",
    "!pip install 'glycowork[ml]'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "GBctW7tjN5A1",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 36
    },
    "id": "GBctW7tjN5A1",
    "outputId": "fce003b8-505e-456d-ccca-c84efd6f913c"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'cuda'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import torch\n",
    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
    "device"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88d4d337-aa5c-4234-bd46-53ff13affddc",
   "metadata": {
    "id": "88d4d337-aa5c-4234-bd46-53ff13affddc"
   },
   "source": [
    "# Fetching ESM Representations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b347be20-c523-4ab3-b400-ba4a6217e07a",
   "metadata": {
    "id": "b347be20-c523-4ab3-b400-ba4a6217e07a"
   },
   "source": [
    "Facebook `fair esm` provides protein embeddings for amino acid sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5a4da62-4569-40a3-8596-0713b7b8e193",
   "metadata": {
    "id": "d5a4da62-4569-40a3-8596-0713b7b8e193"
   },
   "outputs": [],
   "source": [
    "# Run the following on GPU\n",
    "\n",
    "# from google.colab import drive\n",
    "# drive.mount('/content/drive')\n",
    "\n",
    "# !pip install glycowork\n",
    "# !pip install 'glycowork[ml]'\n",
    "# !pip install transformers\n",
    "\n",
    "import torch\n",
    "from transformers import AutoTokenizer, AutoModel\n",
    "from glycowork.glycan_data.loader import glycan_binding\n",
    "import pickle\n",
    "\n",
    "# model_checkpoint = \"facebook/esm2_t33_650M_UR50D\"\n",
    "model_checkpoint = \"facebook/esm1b_t33_650M_UR50S\"\n",
    "\n",
    "tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)\n",
    "model = AutoModel.from_pretrained(model_checkpoint)\n",
    "\n",
    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
    "model.to(device)\n",
    "\n",
    "# Fetch all lectin amino acid sequences (1392 lectins from glycan_binding)\n",
    "all_prots = list(set(glycan_binding.target.tolist()))\n",
    "batch_size = 10\n",
    "\n",
    "# Path to save the pickle file\n",
    "filename = fs + 'esm1b_embeddings_full.pkl'\n",
    "\n",
    "# Function to process a batch of proteins\n",
    "def process_batch(proteins):\n",
    "    inputs = tokenizer(proteins, padding=True, max_length=1000, truncation=True, return_tensors=\"pt\")\n",
    "    inputs = {key: tensor.to(device) for key, tensor in inputs.items()}\n",
    "\n",
    "    with torch.no_grad():\n",
    "        outputs = model(**inputs)\n",
    "    last_hidden_states = outputs.last_hidden_state\n",
    "\n",
    "    return {prot: last_hidden_states[i, 1:len(prot) + 1].mean(0).cpu().numpy().tolist() for i, prot in enumerate(proteins)}\n",
    "\n",
    "# Iterate over the proteins in batches and save the results\n",
    "with open(filename, 'ab') as file:  # 'ab' for appending in binary mode\n",
    "    for i in range(0, len(all_prots), batch_size):\n",
    "        batch_prots = all_prots[i:i + batch_size]\n",
    "        prot_dic = process_batch(batch_prots)\n",
    "        pickle.dump(prot_dic, file)\n",
    "\n",
    "print(\"All data processed and saved successfully.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f436b3b1-e141-4dea-a071-9a33374650be",
   "metadata": {
    "id": "f436b3b1-e141-4dea-a071-9a33374650be"
   },
   "source": [
    "After executing the following code, `data/esm1b_embeddings_full.pkl` and `data/esm2_embeddings_full.pkl` are generated."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ebc1dab-bad9-48d5-8287-79e3bfb8e120",
   "metadata": {
    "id": "5ebc1dab-bad9-48d5-8287-79e3bfb8e120"
   },
   "source": [
    "# Loading Saved ESM Embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "19570d96-390e-40a8-863b-0dce0b40fef6",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "19570d96-390e-40a8-863b-0dce0b40fef6",
    "outputId": "44b48100-08c2-4c98-f41b-373f9a69ed3d"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded data for 1392 proteins.\n"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "\n",
    "# Path to the pickle file\n",
    "filename = fs + 'esm1b_embeddings_full.pkl'\n",
    "\n",
    "# Initialize an empty dictionary to hold all the data\n",
    "prot_dic = {}\n",
    "\n",
    "# Open the pickle file and read from it\n",
    "with open(filename, 'rb') as file:\n",
    "    while True:\n",
    "        try:\n",
    "            # Load the data from the file and update the main dictionary\n",
    "            data = pickle.load(file)\n",
    "            prot_dic.update(data)\n",
    "        except EOFError:\n",
    "            # End of file reached\n",
    "            break\n",
    "\n",
    "# Check the loaded data\n",
    "print(f\"Loaded data for {len(prot_dic)} proteins.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc8d116d-cfec-410c-8764-7ffbb9d58101",
   "metadata": {
    "id": "bc8d116d-cfec-410c-8764-7ffbb9d58101"
   },
   "source": [
    "# Conversion of IUPAC Glycan Motif String Representation to Graph Object"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cad98f61-8cd6-49c0-ba3a-e0dfcae92579",
   "metadata": {
    "id": "cad98f61-8cd6-49c0-ba3a-e0dfcae92579"
   },
   "source": [
    "# Training Lectin Oracle Model from Scratch"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ee0a36a-2ec1-41a9-8452-e78226f3eb7f",
   "metadata": {
    "id": "3ee0a36a-2ec1-41a9-8452-e78226f3eb7f"
   },
   "source": [
    "## Create Torch DataLoaders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "s8rOBfkQLuj9",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "s8rOBfkQLuj9",
    "outputId": "616f5560-ee26-4c9d-c86d-2bed0bcc1f2f"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "87\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "927"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from glycowork.glycan_data.loader import glycan_binding\n",
    "from glycowork.motif.graph import glycan_to_nxGraph\n",
    "from glycowork.motif.processing import get_lib, expand_lib\n",
    "\n",
    "\n",
    "lib = get_lib(glycan_binding.columns.unique().tolist()[:-2])\n",
    "lib = expand_lib(lib, ['GlcNAcOS', 'GalOS', 'HexNAc'])\n",
    "print(len(lib))\n",
    "# generate a dictionary to look up graph representation of glycan\n",
    "\n",
    "glycan_graph_dic = {glyc_string: glycan_to_nxGraph(glyc_string, libr=lib) for glyc_string in glycan_binding.columns.unique().tolist()[:-2]}\n",
    "len(glycan_graph_dic)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6dff54ac-7b0b-4f07-a4f1-b58db64c5e52",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 223
    },
    "id": "6dff54ac-7b0b-4f07-a4f1-b58db64c5e52",
    "outputId": "d9a70731-3b42-42e4-8a3f-4736af037236"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded data for 1393 proteins.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "      <th>glycan</th>\n",
       "      <th>binding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AADSIPSISPTGIITPTPTQSGMVSNCNKFYDVHSNDGCSAIASSQ...</td>\n",
       "      <td>Fuc(a1-2)Gal</td>\n",
       "      <td>0.293462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AADSIPSISPTGIITPTPTQSGMVSNCNKFYDVHSNDGCSAIASSQ...</td>\n",
       "      <td>Fuc(a1-2)Gal(b1-3)GalNAc</td>\n",
       "      <td>-1.316793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AADSIPSISPTGIITPTPTQSGMVSNCNKFYDVHSNDGCSAIASSQ...</td>\n",
       "      <td>Fuc(a1-2)Gal(b1-3)GalNAc(a1-3)[Fuc(a1-2)]Gal(b...</td>\n",
       "      <td>-0.860744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AADSIPSISPTGIITPTPTQSGMVSNCNKFYDVHSNDGCSAIASSQ...</td>\n",
       "      <td>Fuc(a1-2)Gal(b1-3)GalNAc(a1-3)[Fuc(a1-2)]Gal(b...</td>\n",
       "      <td>-1.211838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AADSIPSISPTGIITPTPTQSGMVSNCNKFYDVHSNDGCSAIASSQ...</td>\n",
       "      <td>Fuc(a1-2)Gal(b1-3)GalNAc(b1-3)Gal</td>\n",
       "      <td>-0.335253</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 seq  \\\n",
       "0  AADSIPSISPTGIITPTPTQSGMVSNCNKFYDVHSNDGCSAIASSQ...   \n",
       "1  AADSIPSISPTGIITPTPTQSGMVSNCNKFYDVHSNDGCSAIASSQ...   \n",
       "2  AADSIPSISPTGIITPTPTQSGMVSNCNKFYDVHSNDGCSAIASSQ...   \n",
       "3  AADSIPSISPTGIITPTPTQSGMVSNCNKFYDVHSNDGCSAIASSQ...   \n",
       "4  AADSIPSISPTGIITPTPTQSGMVSNCNKFYDVHSNDGCSAIASSQ...   \n",
       "\n",
       "                                              glycan   binding  \n",
       "0                                       Fuc(a1-2)Gal  0.293462  \n",
       "1                           Fuc(a1-2)Gal(b1-3)GalNAc -1.316793  \n",
       "2  Fuc(a1-2)Gal(b1-3)GalNAc(a1-3)[Fuc(a1-2)]Gal(b... -0.860744  \n",
       "3  Fuc(a1-2)Gal(b1-3)GalNAc(a1-3)[Fuc(a1-2)]Gal(b... -1.211838  \n",
       "4                  Fuc(a1-2)Gal(b1-3)GalNAc(b1-3)Gal -0.335253  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pickle\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "#retrieve representations for protein sequences\n",
    "filename = fs + 'stored_protein_embeddings.pkl'\n",
    "\n",
    "old_dic = {}\n",
    "with open(filename, 'rb') as file:\n",
    "    old_dic = pickle.load(file)\n",
    "\n",
    "print(f\"Loaded data for {len(old_dic)} proteins.\")\n",
    "\n",
    "#converts data into format (protein, glycan, binding)\n",
    "def generate_pair_data(df):\n",
    "  \"\"\"creates 3-tuples of lectin, glycan, and the corresponding binding value measured in relative fluorescence units\"\"\"\n",
    "  out = []\n",
    "  for k in range(len(df)):\n",
    "    for j in range(df.shape[1]-1):\n",
    "      if ~np.isnan(df.iloc[k,j]) and np.isfinite(df.iloc[k,j]):\n",
    "        out.append((df.loc[k,'target'], df.columns.values.tolist()[j], df.iloc[k,j]))\n",
    "  return out\n",
    "\n",
    "data = generate_pair_data(glycan_binding.iloc[:, :-1])\n",
    "df = pd.DataFrame(data, columns = ['seq', 'glycan', 'binding'])\n",
    "df.to_csv(fs + 'all_binding_pairs.csv', index=False)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "609daf07-c8e9-4c6b-bf3a-23fec6b4d8d8",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 432
    },
    "id": "609daf07-c8e9-4c6b-bf3a-23fec6b4d8d8",
    "outputId": "9b74a7d9-e160-4870-d935-d76b563a01bf"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                 seq  \\\n",
      "0  AADSIPSISPTGIITPTPTQSGMVSNCNKFYDVHSNDGCSAIASSQ...   \n",
      "1  AADSIPSISPTGIITPTPTQSGMVSNCNKFYDVHSNDGCSAIASSQ...   \n",
      "2  AADSIPSISPTGIITPTPTQSGMVSNCNKFYDVHSNDGCSAIASSQ...   \n",
      "3  AADSIPSISPTGIITPTPTQSGMVSNCNKFYDVHSNDGCSAIASSQ...   \n",
      "4  AADSIPSISPTGIITPTPTQSGMVSNCNKFYDVHSNDGCSAIASSQ...   \n",
      "\n",
      "                                              glycan   binding  \n",
      "0                                       Fuc(a1-2)Gal  0.293462  \n",
      "1                           Fuc(a1-2)Gal(b1-3)GalNAc -1.316793  \n",
      "2  Fuc(a1-2)Gal(b1-3)GalNAc(a1-3)[Fuc(a1-2)]Gal(b... -0.860744  \n",
      "3  Fuc(a1-2)Gal(b1-3)GalNAc(a1-3)[Fuc(a1-2)]Gal(b... -1.211838  \n",
      "4                  Fuc(a1-2)Gal(b1-3)GalNAc(b1-3)Gal -0.335253  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "      <th>glycan</th>\n",
       "      <th>binding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>55048</th>\n",
       "      <td>[0.04727237671613693, 0.25703755021095276, 0.0...</td>\n",
       "      <td>(0, 1, 2)</td>\n",
       "      <td>-0.143531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>511556</th>\n",
       "      <td>[0.07094207406044006, 0.2685469388961792, 0.02...</td>\n",
       "      <td>(0, 1, 2)</td>\n",
       "      <td>-0.160575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76083</th>\n",
       "      <td>[0.03967776522040367, 0.25796017050743103, 0.0...</td>\n",
       "      <td>(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...</td>\n",
       "      <td>-0.268080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>557630</th>\n",
       "      <td>[0.10139428824186325, 0.2674408257007599, 0.03...</td>\n",
       "      <td>(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10)</td>\n",
       "      <td>-0.282433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207864</th>\n",
       "      <td>[0.016354024410247803, 0.233301043510437, 0.12...</td>\n",
       "      <td>(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...</td>\n",
       "      <td>-0.324050</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      seq  \\\n",
       "55048   [0.04727237671613693, 0.25703755021095276, 0.0...   \n",
       "511556  [0.07094207406044006, 0.2685469388961792, 0.02...   \n",
       "76083   [0.03967776522040367, 0.25796017050743103, 0.0...   \n",
       "557630  [0.10139428824186325, 0.2674408257007599, 0.03...   \n",
       "207864  [0.016354024410247803, 0.233301043510437, 0.12...   \n",
       "\n",
       "                                                   glycan   binding  \n",
       "55048                                           (0, 1, 2) -0.143531  \n",
       "511556                                          (0, 1, 2) -0.160575  \n",
       "76083   (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,... -0.268080  \n",
       "557630                 (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10) -0.282433  \n",
       "207864  (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,... -0.324050  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "df = pd.read_csv(fs + 'all_binding_pairs.csv')\n",
    "print(df.head())\n",
    "\n",
    "df['seq'] = df['seq'].map(old_dic)\n",
    "\n",
    "df['glycan'] = df['glycan'].map(glycan_graph_dic)\n",
    "\n",
    "df_train, df_test = train_test_split(df, test_size = 0.1, random_state = 1, shuffle = True)\n",
    "df_train.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "LVkomGo_SSmF",
   "metadata": {
    "id": "LVkomGo_SSmF"
   },
   "outputs": [],
   "source": [
    "from sklearn.model_selection import KFold, train_test_split\n",
    "from torch_geometric.utils.convert import from_networkx\n",
    "from torch_geometric.loader import DataLoader\n",
    "from torch.utils.data import Subset\n",
    "\n",
    "def create_folds(data, n_splits=10):\n",
    "\n",
    "    kf = KFold(n_splits=n_splits, shuffle=True, random_state=42)\n",
    "\n",
    "    for train_idx, val_idx in kf.split(data):\n",
    "        yield train_idx, val_idx\n",
    "\n",
    "def dataset_to_graphs(glycan_graphs, labels, libr=None, label_type = torch.float):\n",
    "\n",
    "    if libr is None:\n",
    "      libr = lib\n",
    "    data = [from_networkx(k) for k in glycan_graphs]\n",
    "\n",
    "    for data_obj, label in zip(data, labels):\n",
    "        data_obj.y = torch.tensor(label, dtype = label_type)\n",
    "\n",
    "    return data\n",
    "\n",
    "\n",
    "def dataset_to_dataloader(glycan_graphs, labels, libr=lib, batch_size=128, shuffle=True, drop_last = False, extra_feature=None, label_type=torch.float):\n",
    "\n",
    "    if libr is None:\n",
    "      libr = lib\n",
    "\n",
    "    # Converting glycans and labels to PyTorch Geometric Data objects\n",
    "    glycan_graphs = dataset_to_graphs(glycan_graphs, labels, libr = libr, label_type = label_type)\n",
    "\n",
    "    if extra_feature is not None:\n",
    "        for graph, feature in zip(glycan_graphs, extra_feature):\n",
    "            graph.train_idx = torch.tensor(feature, dtype = torch.float)\n",
    "    # Generating the dataloader from the data objects\n",
    "    return DataLoader(glycan_graphs, batch_size = batch_size, shuffle = shuffle, drop_last = drop_last)\n",
    "\n",
    "\n",
    "def get_dataloaders(dataset, train_idx, val_idx, batch_size=128):\n",
    "    \"\"\"\n",
    "    dataset (pd.DataFrame)\n",
    "    train_idx: array of indices to use as examples for training the model during cross validation\n",
    "    val_idx: array of indices to use as tests for validating the model during cross validation\n",
    "    batch_size: size of minibatches\n",
    "    \"\"\"\n",
    "    train_subs = Subset(dataset, train_idx)\n",
    "    val_subs = Subset(dataset, val_idx)\n",
    "\n",
    "    train_loader = dataset_to_dataloader(train_subs.dataset.glycan.tolist(), train_subs.dataset.binding.tolist(), libr=lib, batch_size=128, shuffle=True, extra_feature=train_subs.dataset.seq.tolist(), label_type= torch.float)\n",
    "    val_loader = dataset_to_dataloader(val_subs.dataset.glycan.tolist(), val_subs.dataset.binding.tolist(), libr=lib, batch_size=128, shuffle=False, extra_feature=val_subs.dataset.seq.tolist(), label_type= torch.float)\n",
    "\n",
    "    return {'train': train_loader, 'val': val_loader}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58c283ed-d0aa-4e94-80a0-36302d2119db",
   "metadata": {
    "id": "58c283ed-d0aa-4e94-80a0-36302d2119db"
   },
   "source": [
    "##  LectinOracle Model Architecture Specification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d09dcfaf-ba47-4196-a4b0-1d327d5d7feb",
   "metadata": {
    "id": "d09dcfaf-ba47-4196-a4b0-1d327d5d7feb"
   },
   "outputs": [],
   "source": [
    "#defining LectinOracle\n",
    "import torch.nn as nn\n",
    "from torch_geometric.nn import TopKPooling, GraphConv\n",
    "from torch_geometric.nn import global_mean_pool as gap, global_max_pool as gmp\n",
    "\n",
    "def sigmoid_range(x, low, high):\n",
    "    \"Sigmoid function with range `(low, high)`\"\n",
    "    return torch.sigmoid(x) * (high - low) + low\n",
    "\n",
    "class SigmoidRange(nn.Module):\n",
    "    \"Sigmoid module with range `(low, x_max)`\"\n",
    "    def __init__(self, low, high):\n",
    "      super(SigmoidRange, self).__init__()\n",
    "      self.low, self.high = low,high\n",
    "    def forward(self, x): return sigmoid_range(x, self.low, self.high)\n",
    "\n",
    "class LectinOracle(nn.Module):\n",
    "    def __init__(self, input_size_glyco, hidden_size, num_classes, data_min,\n",
    "               data_max, input_size_prot = 1280, n_layers = 1):\n",
    "        super(LectinOracle,self).__init__()\n",
    "        self.input_size_prot = input_size_prot # 1280 as defined by ESM1b\n",
    "        self.input_size_glyco = input_size_glyco # 84 as defined by the glycan library len(lib)+1\n",
    "        self.hidden_size = hidden_size\n",
    "        self.num_classes = num_classes\n",
    "        self.n_layers = n_layers\n",
    "\n",
    "        # these are the node embeddings for the glycan graph\n",
    "        self.item_embedding = torch.nn.Embedding(num_embeddings = self.input_size_glyco, embedding_dim = self.hidden_size)\n",
    "\n",
    "        # doesn't seem like this line is used\n",
    "        self.glyco_encoder = nn.Embedding(self.input_size_glyco, self.hidden_size, padding_idx = self.input_size_glyco - 1)\n",
    "\n",
    "        # 3 GCN layers with Pooling\n",
    "        self.conv1 = GraphConv(self.hidden_size, self.hidden_size)\n",
    "        self.pool1 = TopKPooling(self.hidden_size, ratio = 0.8)\n",
    "        self.conv2 = GraphConv(self.hidden_size, self.hidden_size)\n",
    "        self.pool2 = TopKPooling(self.hidden_size, ratio = 0.8)\n",
    "        self.conv3 = GraphConv(self.hidden_size, self.hidden_size)\n",
    "        self.pool3 = TopKPooling(self.hidden_size, ratio = 0.8)\n",
    "\n",
    "\n",
    "        # Layers related to ESM1b\n",
    "        self.prot_encoder1 = nn.Linear(self.input_size_prot, 400)\n",
    "        self.prot_encoder2 = nn.Linear(400, 128)\n",
    "        self.dp_prot1 = nn.Dropout(0.2)\n",
    "        self.dp_prot2 = nn.Dropout(0.1)\n",
    "        self.act_prot1 = torch.nn.LeakyReLU()\n",
    "        self.act_prot2 = torch.nn.LeakyReLU()\n",
    "        self.bn_prot1 = nn.BatchNorm1d(400)\n",
    "        self.bn_prot2 = nn.BatchNorm1d(128)\n",
    "\n",
    "        # Fully connected\n",
    "\n",
    "        self.dp1 = nn.Dropout(0.5)\n",
    "        self.fc1 = nn.Linear(128+2*self.hidden_size, int(np.round(self.hidden_size/2)))\n",
    "        self.fc2 = nn.Linear(int(np.round(self.hidden_size/2)), num_classes)\n",
    "        self.bn1 = nn.BatchNorm1d(int(np.round(self.hidden_size/2)))\n",
    "        self.act1 = torch.nn.LeakyReLU()\n",
    "\n",
    "        self.sigmoid = SigmoidRange(data_min, data_max)\n",
    "\n",
    "\n",
    "    def forward(self, prot, nodes, edge_index, batch, inference = False):\n",
    "        # 1280 -> 400 -> 128  (two hidden layers, feedforward with dropout 0.2 on the first, then dropout 0.1 on the second)\n",
    "        # batch norm applied after Leaky ReLU activation??\n",
    "        embedded_prot = self.bn_prot1(self.act_prot1(self.dp_prot1(self.prot_encoder1(prot))))\n",
    "        embedded_prot = self.bn_prot2(self.act_prot2(self.dp_prot2(self.prot_encoder2(embedded_prot))))\n",
    "\n",
    "        # x is the node vector which originaly has len(lib) size but we use torch.nn.Embedding which\n",
    "        x = self.item_embedding(nodes)\n",
    "        x = x.squeeze(1)\n",
    "\n",
    "        x = F.leaky_relu(self.conv1(x, edge_index))\n",
    "\n",
    "        x, edge_index, _, batch, _, _= self.pool1(x, edge_index, None, batch)\n",
    "        x1 = torch.cat([gmp(x, batch), gap(x, batch)], dim = 1) # global mean pool, global max pool\n",
    "\n",
    "        x = F.leaky_relu(self.conv2(x, edge_index))\n",
    "\n",
    "        x, edge_index, _, batch, _, _ = self.pool2(x, edge_index, None, batch)\n",
    "        x2 = torch.cat([gmp(x, batch), gap(x, batch)], dim = 1)\n",
    "\n",
    "        x = F.leaky_relu(self.conv3(x, edge_index))\n",
    "\n",
    "        x, edge_index, _, batch, _, _ = self.pool3(x, edge_index, None, batch)\n",
    "        x3 = torch.cat([gmp(x, batch), gap(x, batch)], dim = 1)\n",
    "\n",
    "        x = x1 + x2 + x3 # is this a bit like ResNet direct connections?\n",
    "\n",
    "        h_n = torch.cat((embedded_prot, x), 1) # concatenated representation of lectin, glycan before fed into feed forward layers\n",
    "\n",
    "        h_n = self.bn1(self.act1(self.fc1(h_n)))\n",
    "\n",
    "        # some kind of ensemble averaging going on?\n",
    "        #1\n",
    "        x1 = self.fc2(self.dp1(h_n))\n",
    "        #2\n",
    "        x2 = self.fc2(self.dp1(h_n))\n",
    "        #3\n",
    "        x3 = self.fc2(self.dp1(h_n))\n",
    "        #4\n",
    "        x4 = self.fc2(self.dp1(h_n))\n",
    "        #5\n",
    "        x5 = self.fc2(self.dp1(h_n))\n",
    "        #6\n",
    "        x6 = self.fc2(self.dp1(h_n))\n",
    "        #7\n",
    "        x7 = self.fc2(self.dp1(h_n))\n",
    "        #8\n",
    "        x8 = self.fc2(self.dp1(h_n))\n",
    "\n",
    "        out =  self.sigmoid(torch.mean(torch.stack([x1, x2, x3, x4, x5, x6, x7, x8]), dim = 0))\n",
    "\n",
    "        if inference:\n",
    "          return out, embedded_prot, x\n",
    "        else:\n",
    "          return out"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8c61fcc-becf-4ee4-98a3-e3605806122f",
   "metadata": {
    "id": "e8c61fcc-becf-4ee4-98a3-e3605806122f"
   },
   "source": [
    "## Model Fitting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69b5579e-963c-4c7f-bac1-e38b3bcb8a7a",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 425
    },
    "id": "69b5579e-963c-4c7f-bac1-e38b3bcb8a7a",
    "outputId": "2499d855-1ee3-42fd-a553-fc1ae38a0cbc"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[     0      1      2 ... 508179 508180 508181] [     6      7     30 ... 508129 508149 508171]\n"
     ]
    }
   ],
   "source": [
    "from glycowork.ml.model_training import train_model\n",
    "\n",
    "def init_weights(m):\n",
    "    if type(m) == nn.Linear:\n",
    "        torch.nn.init.xavier_uniform(m.weight)\n",
    "\n",
    "\n",
    "data_min = glycan_binding.iloc[:, :-2].min().min()\n",
    "data_max = glycan_binding.iloc[:, :-2].max().max()\n",
    "\n",
    "# Setting up 10 fold cross validation\n",
    "\n",
    "n_splits = 10\n",
    "for train_idx, val_idx in create_folds(df_train, n_splits):\n",
    "    print(train_idx, val_idx)\n",
    "    dataloaders = get_dataloaders(df_train, train_idx, val_idx)\n",
    "    model = LectinOracle(input_size_prot = 1280, input_size_glyco = len(lib) + 1, hidden_size = 128, num_classes = 1, data_min = data_min, data_max = data_max)\n",
    "    model.apply(init_weights)\n",
    "    model.cuda()\n",
    "    optimizer = torch.optim.Adam(model.parameters(), lr = 0.0005, weight_decay = 0.0)\n",
    "    scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, 80)\n",
    "    criterion = nn.MSELoss().cuda()\n",
    "    model_ft = train_model(model, dataloaders, optimizer=optimizer, scheduler=scheduler, criterion=criterion, num_epochs = 100, patience = 20, mode = 'regression')\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2eacce16-68bf-4c89-b791-5917acccb95e",
   "metadata": {
    "id": "2eacce16-68bf-4c89-b791-5917acccb95e",
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Using Pretrained LectinOracle Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dc01421-1934-4ff8-abc0-bcc6c88518ea",
   "metadata": {
    "id": "5dc01421-1934-4ff8-abc0-bcc6c88518ea"
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import torch\n",
    "from glycowork.glycan_data.loader import lib, unwrap, build_custom_df, df_glycan, glycan_binding\n",
    "from glycowork.ml.processing import dataset_to_dataloader\n",
    "from glycowork.motif.tokenization import prot_to_coded\n",
    "from glycowork.ml.models import prep_model\n",
    "\n",
    "# Revised helper function that will enable us to retrieve both the prot representation of the model and the binding prediction\n",
    "def get_multi_pred(prot, glycans, model, prot_dic,\n",
    "                   background_correction = False, correction_df = None,\n",
    "                   batch_size = 128, libr = None, flex = False, mode = 'rep'):\n",
    "  \"\"\"Inner function to actually get predictions for lectin-glycan binding from LectinOracle-type model\\n\n",
    "  | Arguments:\n",
    "  | :-\n",
    "  | prot (string): protein amino acid sequence\n",
    "  | glycans (list): list of glycans in IUPACcondensed\n",
    "  | model (PyTorch object): trained LectinOracle-type model\n",
    "  | prot_dic (dictionary): dictionary of type protein sequence:ESM1b representation\n",
    "  | background_correction (bool): whether to correct predictions for background; default:False\n",
    "  | correction_df (dataframe): background prediction for (ideally) all provided glycans; default:None\n",
    "  | batch_size (int): change to batch_size used during training; default:128\n",
    "  | libr (dict): dictionary of form glycoletter:index\n",
    "  | flex (bool): depends on whether you use LectinOracle (False) or LectinOracle_flex (True); default:False\\n\n",
    "  | Returns:\n",
    "  | :-\n",
    "  | Returns dataframe of glycan sequences and predicted binding to prot\n",
    "  \"\"\"\n",
    "  if libr is None:\n",
    "      libr = lib\n",
    "  # Preparing dataset for PyTorch\n",
    "  if flex:\n",
    "      prot = prot_to_coded([prot])\n",
    "      feature = prot * len(glycans)\n",
    "  else:\n",
    "      rep = prot_dic.get(prot, \"new protein, no stored embedding\")\n",
    "      feature = [rep] * len(glycans)\n",
    "  train_loader = dataset_to_dataloader(glycans, [0.99]*len(glycans),\n",
    "                                         libr = libr, batch_size = batch_size,\n",
    "                                         shuffle = False, extra_feature = feature)\n",
    "  model = model.eval()\n",
    "  res = []\n",
    "  # Get predictions for each mini-batch\n",
    "  for k in train_loader:\n",
    "    x, y, edge_index, prot, batch = k.labels, k.y, k.edge_index, k.train_idx, k.batch\n",
    "    x, y, edge_index, prot, batch = x.to(device), y.to(device), edge_index.to(device), prot.view(max(batch) + 1, -1).float().to(device), batch.to(device)\n",
    "    pred, reps, _ = model(prot, x, edge_index, batch, inference=True)\n",
    "    if mode == 'rep':\n",
    "      res.append(reps)\n",
    "    else:\n",
    "      res.append(pred)\n",
    "  res = unwrap([res[k].detach().cpu().numpy() for k in range(len(res))])\n",
    "\n",
    "  return res\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58d44cc1-046c-4da7-9ef6-80326efc3e0e",
   "metadata": {
    "id": "58d44cc1-046c-4da7-9ef6-80326efc3e0e",
    "outputId": "9253ac9a-de7b-42e6-89af-1022e51e9152"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2366\n",
      "Number of parameters: 983089\n"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "from torch_geometric.nn import GraphConv\n",
    "from torch_geometric.nn import global_mean_pool as gap\n",
    "\n",
    "\n",
    "def sigmoid_range(x, low, high):\n",
    "    \"Sigmoid function with range `(low, high)`\"\n",
    "    return torch.sigmoid(x) * (high - low) + low\n",
    "\n",
    "\n",
    "class SigmoidRange(torch.nn.Module):\n",
    "    \"Sigmoid module with range `(low, x_max)`\"\n",
    "\n",
    "    def __init__(self, low, high):\n",
    "      super(SigmoidRange, self).__init__()\n",
    "      self.low, self.high = low, high\n",
    "\n",
    "    def forward(self, x):\n",
    "        return sigmoid_range(x, self.low, self.high)\n",
    "\n",
    "\n",
    "class LectinOracle(nn.Module):\n",
    "    def __init__(self, input_size_glyco, hidden_size, num_classes, data_min,\n",
    "               data_max, input_size_prot = 1280, n_layers = 1):\n",
    "\n",
    "        super(LectinOracle,self).__init__()\n",
    "        self.input_size_prot = input_size_prot # 1280\n",
    "        self.input_size_glyco = input_size_glyco # 84\n",
    "        self.hidden_size = hidden_size # 128\n",
    "        self.num_classes = num_classes # 1\n",
    "        self.n_layers = n_layers # 1\n",
    "\n",
    "        self.conv1 = GraphConv(self.hidden_size, self.hidden_size)\n",
    "        self.pool1 = TopKPooling(self.hidden_size, ratio = 0.8)\n",
    "        self.conv2 = GraphConv(self.hidden_size, self.hidden_size)\n",
    "        self.pool2 = TopKPooling(self.hidden_size, ratio = 0.8)\n",
    "        self.conv3 = GraphConv(self.hidden_size, self.hidden_size)\n",
    "        self.pool3 = TopKPooling(self.hidden_size, ratio = 0.8)\n",
    "\n",
    "        # Glycan node embedding 84 -> 128\n",
    "        self.item_embedding = torch.nn.Embedding(num_embeddings = self.input_size_glyco, embedding_dim = self.hidden_size, padding_idx = self.input_size_glyco - 1)\n",
    "\n",
    "        # ESM1b 1280 -> 400 -> 128\n",
    "        self.prot_encoder1 = nn.Linear(self.input_size_prot, 400)\n",
    "        self.prot_encoder2 = nn.Linear(400, 128)\n",
    "\n",
    "        self.dp1 = nn.Dropout(0.5)\n",
    "        self.dp_prot1 = nn.Dropout(0.2)\n",
    "        self.dp_prot2 = nn.Dropout(0.1)\n",
    "        self.fc1 = nn.Linear(128+2*self.hidden_size, int(np.round(self.hidden_size/2)))\n",
    "        self.fc2 = nn.Linear(int(np.round(self.hidden_size/2)), num_classes)\n",
    "        self.bn1 = nn.BatchNorm1d(int(np.round(self.hidden_size/2)))\n",
    "        self.bn_prot1 = nn.BatchNorm1d(400)\n",
    "        self.bn_prot2 = nn.BatchNorm1d(128)\n",
    "        self.act1 = torch.nn.LeakyReLU()\n",
    "        self.act_prot1 = torch.nn.LeakyReLU()\n",
    "        self.act_prot2 = torch.nn.LeakyReLU()\n",
    "        self.sigmoid = SigmoidRange(data_min, data_max)\n",
    "\n",
    "\n",
    "    def forward(self, prot, nodes, edge_index, batch, inference = False):\n",
    "\n",
    "        # ESM1b 1280 -> 400 -> 128\n",
    "        embedded_prot = self.bn_prot1(self.act_prot1(self.dp_prot1(self.prot_encoder1(prot))))\n",
    "        embedded_prot = self.bn_prot2(self.act_prot2(self.dp_prot2(self.prot_encoder2(embedded_prot))))\n",
    "\n",
    "        # glycan graph\n",
    "        x = self.item_embedding(nodes)\n",
    "        x = x.squeeze(1)\n",
    "\n",
    "        x = F.leaky_relu(self.conv1(x, edge_index))\n",
    "\n",
    "        x, edge_index, _, batch, _, _= self.pool1(x, edge_index, None, batch)\n",
    "        x1 = torch.cat([gmp(x, batch), gap(x, batch)], dim = 1)\n",
    "\n",
    "        x = F.leaky_relu(self.conv2(x, edge_index))\n",
    "\n",
    "        x, edge_index, _, batch, _, _ = self.pool2(x, edge_index, None, batch)\n",
    "        x2 = torch.cat([gmp(x, batch), gap(x, batch)], dim = 1)\n",
    "\n",
    "        x = F.leaky_relu(self.conv3(x, edge_index))\n",
    "\n",
    "        x, edge_index, _, batch, _, _ = self.pool3(x, edge_index, None, batch)\n",
    "        x3 = torch.cat([gmp(x, batch), gap(x, batch)], dim = 1)\n",
    "\n",
    "        # some resnet style architecture going on here? this is not just GCN layered on top of each other...\n",
    "        x = x1 + x2 + x3\n",
    "\n",
    "        # glycan graph embeddings + ESM1 embeddings concactenated this is the start of the fully connected layer\n",
    "        h_n = torch.cat((embedded_prot, x), 1)\n",
    "\n",
    "        h_n = self.bn1(self.act1(self.fc1(h_n)))\n",
    "\n",
    "        # some kind of ensemble method going on here?\n",
    "        #1\n",
    "        x1 = self.fc2(self.dp1(h_n))\n",
    "        #2\n",
    "        x2 = self.fc2(self.dp1(h_n))\n",
    "        #3\n",
    "        x3 = self.fc2(self.dp1(h_n))\n",
    "        #4\n",
    "        x4 = self.fc2(self.dp1(h_n))\n",
    "        #5\n",
    "        x5 = self.fc2(self.dp1(h_n))\n",
    "        #6\n",
    "        x6 = self.fc2(self.dp1(h_n))\n",
    "        #7\n",
    "        x7 = self.fc2(self.dp1(h_n))\n",
    "        #8\n",
    "        x8 = self.fc2(self.dp1(h_n))\n",
    "\n",
    "        out =  self.sigmoid(torch.mean(torch.stack([x1, x2, x3, x4, x5, x6, x7, x8]), dim = 0))\n",
    "\n",
    "        if inference:\n",
    "          return out, embedded_prot, x\n",
    "        else:\n",
    "          return out\n",
    "\n",
    "\n",
    "# Choosing the right computing architecture\n",
    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
    "\n",
    "\n",
    "from glycowork.glycan_data.loader import lib\n",
    "\n",
    "print(len(lib))\n",
    "\n",
    "model = LectinOracle(len(lib), num_classes = 1)\n",
    "model.load_state_dict(torch.load('models/glycowork_lectinoracle_600.pt', map_location = device))\n",
    "model.to(device)\n",
    "\n",
    "num_params = sum(p.numel() for p in model.parameters() if p.requires_grad)\n",
    "print(f\"Number of parameters: {num_params}\")\n",
    "\n",
    "# LeOr_flex = prep_model('LectinOracle_flex', 1, trained = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a40567e-c1e5-42cf-8e1a-3af0fd38d121",
   "metadata": {
    "id": "1a40567e-c1e5-42cf-8e1a-3af0fd38d121",
    "outputId": "9089afb0-aee7-4715-ca4a-ee29733124ac"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "83\n"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "Error(s) in loading state_dict for LectinOracle:\n\tMissing key(s) in state_dict: \"conv1.lin_rel.weight\", \"conv1.lin_rel.bias\", \"conv1.lin_root.weight\", \"pool1.select.weight\", \"conv2.lin_rel.weight\", \"conv2.lin_rel.bias\", \"conv2.lin_root.weight\", \"pool2.select.weight\", \"conv3.lin_rel.weight\", \"conv3.lin_rel.bias\", \"conv3.lin_root.weight\", \"pool3.select.weight\". \n\tUnexpected key(s) in state_dict: \"conv1.lin_l.weight\", \"conv1.lin_l.bias\", \"conv1.lin_r.weight\", \"pool1.weight\", \"conv2.lin_l.weight\", \"conv2.lin_l.bias\", \"conv2.lin_r.weight\", \"pool2.weight\", \"conv3.lin_l.weight\", \"conv3.lin_l.bias\", \"conv3.lin_r.weight\", \"pool3.weight\". \n\tsize mismatch for item_embedding.weight: copying a param with shape torch.Size([84, 128]) from checkpoint, the shape in current model is torch.Size([83, 128]).\n\tsize mismatch for glyco_encoder.weight: copying a param with shape torch.Size([84, 128]) from checkpoint, the shape in current model is torch.Size([83, 128]).",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[107], line 14\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;28mlen\u001b[39m(lib))\n\u001b[1;32m     11\u001b[0m model \u001b[38;5;241m=\u001b[39m LectinOracle(input_size_prot \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1280\u001b[39m, input_size_glyco \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(lib), hidden_size \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m128\u001b[39m,\n\u001b[1;32m     12\u001b[0m           num_classes \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m, data_min \u001b[38;5;241m=\u001b[39m glycan_binding\u001b[38;5;241m.\u001b[39miloc[:, :\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m2\u001b[39m]\u001b[38;5;241m.\u001b[39mmin()\u001b[38;5;241m.\u001b[39mmin(), data_max \u001b[38;5;241m=\u001b[39m glycan_binding\u001b[38;5;241m.\u001b[39miloc[:, :\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m2\u001b[39m]\u001b[38;5;241m.\u001b[39mmax()\u001b[38;5;241m.\u001b[39mmax())\n\u001b[0;32m---> 14\u001b[0m \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mload_state_dict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtorch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mload\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmodels/LectinOracle_565.pt\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmap_location\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtorch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdevice\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mcpu\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     15\u001b[0m \u001b[38;5;66;03m# model_ft = model_ft.cuda()\u001b[39;00m\n\u001b[1;32m     18\u001b[0m num_params \u001b[38;5;241m=\u001b[39m \u001b[38;5;28msum\u001b[39m(p\u001b[38;5;241m.\u001b[39mnumel() \u001b[38;5;28;01mfor\u001b[39;00m p \u001b[38;5;129;01min\u001b[39;00m model\u001b[38;5;241m.\u001b[39mparameters() \u001b[38;5;28;01mif\u001b[39;00m p\u001b[38;5;241m.\u001b[39mrequires_grad)\n",
      "File \u001b[0;32m~/miniconda3/envs/biopython/lib/python3.9/site-packages/torch/nn/modules/module.py:2152\u001b[0m, in \u001b[0;36mModule.load_state_dict\u001b[0;34m(self, state_dict, strict, assign)\u001b[0m\n\u001b[1;32m   2147\u001b[0m         error_msgs\u001b[38;5;241m.\u001b[39minsert(\n\u001b[1;32m   2148\u001b[0m             \u001b[38;5;241m0\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMissing key(s) in state_dict: \u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m. \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mformat(\n\u001b[1;32m   2149\u001b[0m                 \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m, \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mk\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m k \u001b[38;5;129;01min\u001b[39;00m missing_keys)))\n\u001b[1;32m   2151\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(error_msgs) \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m-> 2152\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mError(s) in loading state_dict for \u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m:\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mformat(\n\u001b[1;32m   2153\u001b[0m                        \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(error_msgs)))\n\u001b[1;32m   2154\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m _IncompatibleKeys(missing_keys, unexpected_keys)\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Error(s) in loading state_dict for LectinOracle:\n\tMissing key(s) in state_dict: \"conv1.lin_rel.weight\", \"conv1.lin_rel.bias\", \"conv1.lin_root.weight\", \"pool1.select.weight\", \"conv2.lin_rel.weight\", \"conv2.lin_rel.bias\", \"conv2.lin_root.weight\", \"pool2.select.weight\", \"conv3.lin_rel.weight\", \"conv3.lin_rel.bias\", \"conv3.lin_root.weight\", \"pool3.select.weight\". \n\tUnexpected key(s) in state_dict: \"conv1.lin_l.weight\", \"conv1.lin_l.bias\", \"conv1.lin_r.weight\", \"pool1.weight\", \"conv2.lin_l.weight\", \"conv2.lin_l.bias\", \"conv2.lin_r.weight\", \"pool2.weight\", \"conv3.lin_l.weight\", \"conv3.lin_l.bias\", \"conv3.lin_r.weight\", \"pool3.weight\". \n\tsize mismatch for item_embedding.weight: copying a param with shape torch.Size([84, 128]) from checkpoint, the shape in current model is torch.Size([83, 128]).\n\tsize mismatch for glyco_encoder.weight: copying a param with shape torch.Size([84, 128]) from checkpoint, the shape in current model is torch.Size([83, 128])."
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Choosing the right computing architecture\n",
    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
    "\n",
    "from glycowork.motif.processing import get_lib, expand_lib\n",
    "\n",
    "lib = get_lib(glycan_binding.columns.unique().tolist()[:-2])\n",
    "lib = expand_lib(lib, ['GlcNAcOS', 'GalOS', 'HexNAc'])\n",
    "\n",
    "print(len(lib))\n",
    "\n",
    "model = LectinOracle(input_size_prot = 1280, input_size_glyco = len(lib), hidden_size = 128,\n",
    "          num_classes = 1, data_min = glycan_binding.iloc[:, :-2].min().min(), data_max = glycan_binding.iloc[:, :-2].max().max())\n",
    "\n",
    "model.load_state_dict(torch.load('models/LectinOracle_565.pt', map_location=torch.device('cpu')))\n",
    "# model_ft = model_ft.cuda()\n",
    "\n",
    "\n",
    "num_params = sum(p.numel() for p in model.parameters() if p.requires_grad)\n",
    "print(f\"Number of parameters: {num_params}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69d36c95-f57d-4ac6-8d6c-98f28be19f58",
   "metadata": {
    "id": "69d36c95-f57d-4ac6-8d6c-98f28be19f58",
    "outputId": "47968523-efd2-4919-de86-22c4e417dd65"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Fuc(a1-3)[Gal(b1-4)]Glc6S-ol',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc6S(b1-4)Gal(b1-4)[Fuc(a1-3)]GlcNAc-ol',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc6S(b1-6)Man(a1-6)Man(b1-4)GlcNAc-ol',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc6S(b1-6)Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc-ol',\n",
       " 'Fuc4S(a1-3)[Gal(b1-4)]GlcNAc(b1-2)Man(a1-6)Man(b1-4)GlcNAc-ol',\n",
       " 'GalNAc(a1-4)GlcA(b1-2)Glc-ol',\n",
       " 'GalNAc(a1-4)GlcA(b1-3)Gal(b1-3)Gal(b1-4)Xyl-ol',\n",
       " 'GalNAc(a1-4)GlcA(b1-3)GlcNAc-ol',\n",
       " 'GalNAc(a1-4)GlcA(b1-4)GlcNAc-ol',\n",
       " 'GlcA(b1-3)Glc-ol']"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# making predictions for human glycan motifs\n",
    "df_species = build_custom_df(df_glycan, 'df_species')\n",
    "glyc = df_species.loc[df_species.Species == 'Homo_sapiens'].glycan.unique().tolist()\n",
    "\n",
    "# fetching glycan motifs associated with pancreatic cancer according to curated df_disease\n",
    "df_disease = build_custom_df(df_glycan, 'df_disease')\n",
    "df_panc = df_disease[df_disease['disease_association'] == 'pancreatic_cancer']\n",
    "biomarkers = df_panc.glycan.unique().tolist()\n",
    "biomarkers[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "458d3aaa-ae50-4b37-8c85-b3cd28fcaa23",
   "metadata": {
    "id": "458d3aaa-ae50-4b37-8c85-b3cd28fcaa23"
   },
   "outputs": [],
   "source": [
    "#getting learned protein represenations from LectinOracle\n",
    "#this takes a long time to execute\n",
    "prot_reps = {}\n",
    "binding_pred = {}\n",
    "for i, seq in enumerate(glycan_binding.target.tolist()[:2]):\n",
    "  if(i%10 == 0):\n",
    "    print(i)\n",
    "  prot_reps[seq] = np.array(get_multi_pred(seq, [glyc[0]], LeOr, old_dic)).squeeze()\n",
    "  binding_pred[seq] = unwrap(get_multi_pred(seq, glyc, LeOr, old_dic, mode='pred'))\n",
    "\n",
    "# pd.DataFrame(prot_reps).T.to_csv('/content/drive/My Drive/LeOr_embeddings.csv')\n",
    "# motif_pred = pd.DataFrame(binding_pred).T\n",
    "# motif_pred.columns = glyc\n",
    "# motif_pred.to_csv('/content/drive/My Drive/human_glycan_motif_predicted.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96766a01-4078-4f5e-8389-c8207752b335",
   "metadata": {
    "id": "96766a01-4078-4f5e-8389-c8207752b335",
    "outputId": "0a2ada08-ea91-4f34-9e4e-27c0e6069587"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2366"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(lib)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ab49a0e-83ed-40e3-8f1c-2337cfaf1f12",
   "metadata": {
    "id": "1ab49a0e-83ed-40e3-8f1c-2337cfaf1f12",
    "outputId": "3f021ad0-63af-4525-c5d2-9f555d2ab22c"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['3-Anhydro-Gal(a1-3)Gal(b1-4)3-Anhydro-Gal(a1-3)Gal4S',\n",
       " '3-Anhydro-Gal(a1-3)Gal4S(b1-4)3-Anhydro-Gal(a1-3)Gal4S',\n",
       " '3-Anhydro-Gal(a1-3)Gal4S(b1-4)3-Anhydro-Gal(a1-3)Gal4S(b1-4)3-Anhydro-Gal(a1-3)Gal4S',\n",
       " '3-Anhydro-Gal(a1-3)Gal4S(b1-4)3-Anhydro-Gal(a1-3)Gal4S(b1-4)3-Anhydro-Gal(a1-3)Gal4S(b1-4)3-Anhydro-Gal(a1-3)Gal4S',\n",
       " '3-Anhydro-Gal(a1-3)Gal4S(b1-4)3-Anhydro-Gal2S(a1-3)Gal4S(b1-4)3-Anhydro-Gal(a1-3)Gal4S',\n",
       " '3dGal(b1-3)[Fuc(a1-4)]Glc',\n",
       " '3dGal(b1-4)Glc',\n",
       " '4d8dNeu5Ac(a2-3)Gal(b1-4)Glc',\n",
       " '4dNeu5Ac(a2-3)Gal(b1-4)Glc',\n",
       " '7dNeu5Ac(a2-3)Gal(b1-4)Glc',\n",
       " '8dNeu5Ac(a2-3)Gal(b1-4)Glc',\n",
       " '9dNeu5Ac(a2-3)Gal(b1-4)Glc',\n",
       " 'Ara(a1-5)Ara(a1-5)Ara(a1-5)Ara(a1-5)Ara(a1-5)Ara',\n",
       " 'Ara(a1-5)Ara(a1-5)Ara(a1-5)Ara(a1-5)Ara(a1-5)Ara(a1-5)Ara',\n",
       " 'Fuc(a1-2)Gal',\n",
       " 'Fuc(a1-2)Gal(b1-3)GalNAc',\n",
       " 'Fuc(a1-2)Gal(b1-3)GalNAc(a1-3)[Fuc(a1-2)]Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-3)GalNAc(a1-3)[Fuc(a1-2)]Gal(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-3)GalNAc(b1-3)Gal',\n",
       " 'Fuc(a1-2)Gal(b1-3)GalNAc(b1-3)Gal(a1-4)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-3)GalNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-3)GalNAc(b1-4)[Neu5Ac(a2-3)]Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-3)GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-3)GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)Gal(b1-3)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-3)GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)Gal(b1-3)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-3)GlcNAc(b1-3)GalNAc',\n",
       " 'Fuc(a1-2)Gal(b1-3)GlcNAc(b1-3)[Fuc(a1-2)Gal(b1-3)GlcNAc(b1-6)]GalNAc',\n",
       " 'Fuc(a1-2)Gal(b1-3)GlcNAc(b1-3)[Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-3)GlcNAc(b1-3)[Fuc(a1-3)]Gal(b1-4)GlcNAc(b1-6)[Fuc(a1-2)Gal(b1-3)GlcNAc(b1-3)]Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-3)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-3)GlcNAc(b1-3)[Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-3)GlcNAc(b1-3)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-3)GlcNAc6S',\n",
       " 'Fuc(a1-2)Gal(b1-3)[Fuc(a1-4)]GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-6)[Fuc(a1-2)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)[Fuc(a1-2)]Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)[Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-4)Fuc(a1-3)]GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-4)Glc6S',\n",
       " 'Fuc(a1-2)Gal(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)GlcNAc(b1-2)Man',\n",
       " 'Fuc(a1-2)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)[Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)GlcNAc(b1-2)[Fuc(a1-2)Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)[Fuc(a1-2)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)GlcNAc(b1-3)GalNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)GlcNAc(b1-3)[Fuc(a1-2)Gal(b1-4)GlcNAc(b1-6)]GalNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)GlcNAc(b1-3)[Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-4)GlcNAc(b1-6)GalNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)GlcNAc6S',\n",
       " 'Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GalNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GalNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GalNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-6)[Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)[Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-4)]Man(a1-3)[Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)GalNAc',\n",
       " 'Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)[Fuc(a1-2)]Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal3S(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)Gal6S(b1-3)GlcNAc',\n",
       " 'Fuc(a1-2)Gal6S(b1-3)GlcNAc6S',\n",
       " 'Fuc(a1-2)Gal6S(b1-4)Glc',\n",
       " 'Fuc(a1-2)Gal6S(b1-4)Glc6S',\n",
       " 'Fuc(a1-2)Gal6S(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)GlcNAc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-3)GalNAc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-3)GalNAc(b1-3)Gal(a1-4)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-3)GlcNAc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-3)GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)[Gal(a1-3)]Gal(b1-3)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-3)GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)[Gal(a1-3)]Gal(b1-3)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-3)GlcNAc(b1-3)GalNAc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-3)GlcNAc(b1-6)GalNAc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-4)GlcNAc(b1-2)Man',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)[Gal(a1-3)]Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)[Gal(a1-3)]Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-4)GlcNAc(b1-3)GalNAc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-4)GlcNAc(b1-3)[Fuc(a1-2)[Gal(a1-3)]Gal(b1-4)GlcNAc(b1-6)]GalNAc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-4)GlcNAc(b1-6)GalNAc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-4)[Fuc(a1-3)]Glc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Fuc(a1-2)[Gal(a1-3)]Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)GalNAc',\n",
       " 'Fuc(a1-2)[Gal(a1-4)]Gal(b1-4)GalNAc',\n",
       " 'Fuc(a1-2)[Gal(a1-4)]Gal(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-3)GalNAc(a1-3)[Fuc(a1-2)]Gal(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-3)GalNAc(b1-3)Gal(a1-4)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-3)GlcNAc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-3)GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-3)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-3)GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-3)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-3)GlcNAc(b1-3)GalNAc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-3)GlcNAc(b1-6)GalNAc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)GlcNAc(b1-2)Man',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[GalNAc(a1-3)[Fuc(a1-2)]Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)GlcNAc(b1-3)GalNAc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)GlcNAc(b1-3)[Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)GlcNAc(b1-6)]GalNAc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)GlcNAc(b1-6)GalNAc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-3)]Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)GalNAc',\n",
       " 'Fuc(a1-2)[GalNAc(a1-4)]Gal(b1-4)GlcNAc',\n",
       " 'Fuc(a1-2)[GalNAc(b1-3)]Gal',\n",
       " 'Fuc(a1-2)[Neu5Ac(bond)]Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-3)GlcNAc',\n",
       " 'Fuc(a1-3)GlcNAc6S',\n",
       " 'Fuc(a1-3)[3dGal(b1-4)]Glc',\n",
       " 'Fuc(a1-3)[3dGal(b1-4)]GlcNAc(b1-3)Gal',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]Glc',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]Glc6S',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-2)Man',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-2)Man(a1-3)[Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-2)Man(a1-3)[Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-2)Man(a1-6)[Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-3)Gal(b1-3)[Fuc(a1-4)]GlcNAc',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]Glc',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-3)GalNAc',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-3)[Fuc(a1-2)]Gal(b1-4)Glc',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-3)[Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-4)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-4)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-4)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-6)GalNAc',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-6)[Gal(b1-3)GlcNAc(b1-3)]Gal(b1-4)Glc',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-6)[Neu5Ac(a2-3)Gal(b1-3)]GalNAc',\n",
       " 'Fuc(a1-3)[Gal(b1-4)]GlcNAc6S',\n",
       " 'Fuc(a1-3)[Gal3S(b1-4)]Glc',\n",
       " 'Fuc(a1-3)[Gal3S(b1-4)]Glc6S',\n",
       " 'Fuc(a1-3)[Gal3S(b1-4)]GlcNAc',\n",
       " 'Fuc(a1-3)[Gal3S(b1-4)]GlcNAc(b1-3)Gal',\n",
       " 'Fuc(a1-3)[Gal3S(b1-4)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-3)[Gal3S(b1-4)]GlcNAc6S',\n",
       " 'Fuc(a1-3)[Gal3S(b1-4)]GlcNAc6S(b1-3)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-3)[Gal6S(b1-4)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Fuc(a1-3)[GalNAc(b1-4)]Gal(b1-4)GlcNAc',\n",
       " 'Fuc(a1-3)[GalNAc(b1-4)]GlcNAc',\n",
       " 'Fuc(a1-3)[GalNAc(b1-4)]GlcNAc6S',\n",
       " 'Fuc(a1-3)[GalNAc3S(b1-4)]GlcNAc',\n",
       " 'Fuc(a1-3)[Neu5Ac(a2-3)]Gal(b1-4)GlcNAc(b1-3)GalNAc',\n",
       " 'Fuc(a1-4)GlcNAc',\n",
       " 'Fuc(a1-6)GlcNAc',\n",
       " 'Fuc(b1-3)GlcNAc',\n",
       " 'Gal(a1-2)Gal',\n",
       " 'Gal(a1-3)Fuc(a1-2)Gal(b1-4)GlcNAc(b1-3)[Gal(a1-3)Fuc(a1-2)Gal(b1-4)GlcNAc(b1-6)]GalNAc',\n",
       " 'Gal(a1-3)Gal',\n",
       " 'Gal(a1-3)Gal(b1-3)Glc',\n",
       " 'Gal(a1-3)Gal(b1-3)GlcNAc',\n",
       " 'Gal(a1-3)Gal(b1-3)GlcNAc(b1-2)Man',\n",
       " 'Gal(a1-3)Gal(b1-3)GlcNAc(b1-2)Man(a1-3)[Gal(a1-3)Gal(b1-3)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(a1-3)Gal(b1-3)GlcNAc(b1-3)GalNAc',\n",
       " 'Gal(a1-3)Gal(b1-3)GlcNAc(b1-6)GalNAc',\n",
       " 'Gal(a1-3)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-3)[Gal(a1-3)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(a1-3)Gal(b1-4)Gal(a1-3)Gal',\n",
       " 'Gal(a1-3)Gal(b1-4)Glc',\n",
       " 'Gal(a1-3)Gal(b1-4)GlcNAc',\n",
       " 'Gal(a1-3)Gal(b1-4)GlcNAc(b1-2)Man',\n",
       " 'Gal(a1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(a1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(a1-3)Gal(b1-4)GlcNAc(b1-3)GalNAc',\n",
       " 'Gal(a1-3)Gal(b1-4)GlcNAc(b1-6)GalNAc',\n",
       " 'Gal(a1-3)Gal(b1-4)[Fuc(a1-3)]GalNAc',\n",
       " 'Gal(a1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Gal(a1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-3)[Gal(a1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(a1-3)GalNAc',\n",
       " 'Gal(a1-3)GlcNAc',\n",
       " 'Gal(a1-3)[Gal(a1-4)]Gal(b1-4)GlcNAc',\n",
       " 'Gal(a1-4)Gal(b1-3)GlcNAc(b1-2)Man(a1-3)[Gal(a1-4)Gal(b1-3)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(a1-4)Gal(b1-4)Gal',\n",
       " 'Gal(a1-4)Gal(b1-4)GalNAc',\n",
       " 'Gal(a1-4)Gal(b1-4)Glc',\n",
       " 'Gal(a1-4)Gal(b1-4)GlcNAc',\n",
       " 'Gal(a1-4)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(a1-4)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(a1-4)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Gal(a1-4)Glc',\n",
       " 'Gal(a1-4)GlcNAc',\n",
       " 'Gal(a1-6)Glc',\n",
       " 'Gal(b1-2)Gal',\n",
       " 'Gal(b1-3)Gal',\n",
       " 'Gal(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'Gal(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-3)GalNAc',\n",
       " 'Gal(b1-3)GalNAc(a1-3)[Fuc(a1-2)]Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)GalNAc(a1-3)[Fuc(a1-2)]Gal(b1-4)GlcNAc',\n",
       " 'Gal(b1-3)GalNAc(b1-3)Gal',\n",
       " 'Gal(b1-3)GalNAc(b1-3)Gal(a1-4)Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)GalNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)GalNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'Gal(b1-3)GalNAc(b1-4)Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)GalNAc(b1-4)Gal3S(b1-4)Glc',\n",
       " 'Gal(b1-3)GalNAc(b1-4)[Neu5Ac(a2-3)]Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)GalNAc(b1-4)[Neu5Gc(a2-3)]Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)GlcNAc',\n",
       " 'Gal(b1-3)GlcNAc(a1-3)Gal(b1-4)GlcNAc',\n",
       " 'Gal(b1-3)GlcNAc(a1-6)Gal(b1-4)GlcNAc',\n",
       " 'Gal(b1-3)GlcNAc(b1-2)Man',\n",
       " 'Gal(b1-3)GlcNAc(b1-2)Man(a1-3)[Gal(b1-3)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-3)GlcNAc(b1-2)Man(a1-3)[Gal(b1-3)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-3)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-4)][Gal(b1-3)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-3)GlcNAc(b1-2)[Gal(b1-3)GlcNAc(b1-6)]Man(a1-6)[Gal(b1-3)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-3)GlcNAc(b1-3)Gal(b1-3)GlcNAc',\n",
       " 'Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)[Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)[Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)[Gal(b1-3)GlcNAc(b1-3)]Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-6)[Gal(b1-3)GlcNAc(b1-3)]Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)GlcNAc(b1-3)GalNAc',\n",
       " 'Gal(b1-3)GlcNAc(b1-3)[Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)GlcNAc(b1-3)[Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-6)]Gal(b1-4)]GlcNAc(b1-6)[Gal(b1-4)]GlcNAc(b1-3)]Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)GlcNAc(b1-3)[Fuc(a1-3)]Gal(b1-4)GlcNAc(b1-6)[Fuc(a1-2)Gal(b1-3)GlcNAc(b1-3)]Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)GlcNAc(b1-6)[Gal(b1-3)GlcNAc(b1-3)]Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)GlcNAc(b1-6)Gal(b1-4)GlcNAc',\n",
       " 'Gal(b1-3)GlcNAc(b1-6)GalNAc',\n",
       " 'Gal(b1-3)GlcNAc(b1-6)[Gal(b1-3)]GalNAc',\n",
       " 'Gal(b1-3)GlcNAc6S',\n",
       " 'Gal(b1-3)[Fuc(a1-4)]GlcNAc',\n",
       " 'Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man',\n",
       " 'Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-3)[Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-3)[Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)Gal(b1-3)[Fuc(a1-4)]GlcNAc',\n",
       " 'Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]Glc',\n",
       " 'Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-6)[Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)]Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)[Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-6)GalNAc',\n",
       " 'Gal(b1-3)[GlcNAc(b1-6)]GalNAc',\n",
       " 'Gal(b1-3)[GlcNAc(b1-6)]GlcNAc',\n",
       " 'Gal(b1-3)[Neu5Ac(a2-6)]GalNAc',\n",
       " 'Gal(b1-3)[Neu5Ac(a2-6)]GalNAc(b1-4)Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)[Neu5Ac(a2-6)]GlcNAc',\n",
       " 'Gal(b1-3)[Neu5Ac(a2-6)]GlcNAc(a1-4)Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)[Neu5Ac(a2-6)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)[Neu5Ac(a2-6)]GlcNAc(b1-4)Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)[Neu5Ac(a2-6)]GlcNAc(b1-6)Gal(b1-4)Glc',\n",
       " 'Gal(b1-3)[Neu5Ac(b2-6)]GalNAc',\n",
       " 'Gal(b1-3)[Neu5Ac(b2-6)]GlcNAc',\n",
       " 'Gal(b1-4)Gal',\n",
       " 'Gal(b1-4)GalNAc(a1-3)[Fuc(a1-2)]Gal(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GalNAc(b1-3)[Fuc(a1-2)]Gal(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)Glc',\n",
       " 'Gal(b1-4)Glc6S',\n",
       " 'Gal(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(a1-6)Gal(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)Gal(b1-4)GlcNAc(b1-2)GalNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)Man',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)Man(a1-3)Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-2)Man(a1-6)][Gal(b1-4)GlcNAc(b1-4)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-2)Man(a1-6)][GlcNAc(b1-4)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-2)][GlcNAc(b1-4)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-4)][Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)Man(a1-6)Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)Man(a1-6)[GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)Man(a1-6)[GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)[Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)[Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)[Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)[Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)][GlcNAc(b1-4)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)[GlcNAc(b1-4)][Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)[GlcNAc(b1-4)][Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)[GlcNAc(b1-4)][Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[GlcNAc(b1-4)][Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[GlcNAc(b1-4)][Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Fuc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-3)GalNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-3)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)GalNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)]GalNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)[Gal(b1-3)]GalNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)GalNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)[Fuc(a1-3)GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)[Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]GalNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-3)[GlcNAc(b1-6)]Gal(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-6)Gal(b1-4)GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-6)Gal(b1-4)GlcNAc(b1-6)GalNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-6)GalNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-6)[Gal(b1-3)]GalNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-6)[Gal(b1-3)]GlcNAc',\n",
       " 'Gal(b1-4)GlcNAc(b1-6)[Gal(b1-4)GlcNAc(b1-3)]GalNAc',\n",
       " 'Gal(b1-4)GlcNAc6P',\n",
       " 'Gal(b1-6)Gal',\n",
       " 'Gal(b1-6)GalNAc',\n",
       " 'Gal3S(b1-3)GalNAc',\n",
       " 'Gal3S(b1-3)GalNAc(b1-4)Gal3S(b1-4)Glc',\n",
       " 'Gal3S(b1-3)GlcNAc',\n",
       " 'Gal3S(b1-3)[Fuc(a1-4)]GalNAc',\n",
       " 'Gal3S(b1-3)[Fuc(a1-4)]GlcNAc',\n",
       " 'Gal3S(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Gal3S(b1-4)Glc',\n",
       " 'Gal3S(b1-4)Glc6S',\n",
       " 'Gal3S(b1-4)GlcNAc',\n",
       " 'Gal3S(b1-4)GlcNAc(b1-3)Gal',\n",
       " 'Gal3S(b1-4)GlcNAc6S',\n",
       " 'Gal3S(b1-4)[Fuc(a1-3)]Glc',\n",
       " 'Gal3S(b1-4)[Fuc(a1-3)]Glc6S',\n",
       " 'Gal3S(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Gal3S(b1-4)[Fuc(a1-3)]GlcNAc6S',\n",
       " 'Gal3S4S(b1-4)GlcNAc',\n",
       " 'Gal3S6S(b1-4)GlcNAc',\n",
       " 'Gal3S6S(b1-4)GlcNAc6S',\n",
       " 'Gal4S(b1-4)GlcNAc',\n",
       " 'Gal4S6S(b1-4)GlcNAc',\n",
       " 'Gal6P(b1-4)GlcNAc',\n",
       " 'Gal6S(b1-3)GlcNAc',\n",
       " 'Gal6S(b1-3)GlcNAc6S',\n",
       " 'Gal6S(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Gal6S(b1-4)Glc',\n",
       " 'Gal6S(b1-4)Glc6S',\n",
       " 'Gal6S(b1-4)GlcNAc',\n",
       " 'Gal6S(b1-4)GlcNAc6S',\n",
       " 'GalNAc(a1-3)Fuc(a1-2)Gal(b1-4)GlcNAc(b1-3)[GalNAc(a1-3)Fuc(a1-2)Gal(b1-4)GlcNAc(b1-6)]GalNAc',\n",
       " 'GalNAc(a1-3)Gal',\n",
       " 'GalNAc(a1-3)Gal(b1-4)Glc',\n",
       " 'GalNAc(a1-3)GalNAc',\n",
       " 'GalNAc(a1-3)GalNAc(a1-3)Gal(a1-4)Gal(b1-4)GlcNAc',\n",
       " 'GalNAc(a1-3)GalNAc(b1-3)Gal(a1-4)Gal(b1-4)Glc',\n",
       " 'GalNAc(a1-3)GalNAc(b1-3)Gal(a1-4)Gal(b1-4)GlcNAc',\n",
       " 'GalNAc(b1-3)Gal',\n",
       " 'GalNAc(b1-3)Gal(a1-4)Gal(b1-4)GlcNAc',\n",
       " 'GalNAc(b1-3)Gal(a1-4)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'GalNAc(b1-3)Gal(a1-6)Gal(b1-4)Glc',\n",
       " 'GalNAc(b1-3)GalNAc',\n",
       " 'GalNAc(b1-3)GlcNAc',\n",
       " 'GalNAc(b1-4)Gal(b1-4)Glc',\n",
       " 'GalNAc(b1-4)Gal3S(b1-4)Glc',\n",
       " 'GalNAc(b1-4)GlcNAc',\n",
       " 'GalNAc(b1-4)GlcNAc(b1-2)Man',\n",
       " 'GalNAc(b1-4)GlcNAc(b1-2)Man(a1-3)[GalNAc(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GalNAc(b1-4)GlcNAc(b1-3)GalNAc(b1-4)GlcNAc',\n",
       " 'GalNAc(b1-4)GlcNAc6S',\n",
       " 'GalNAc(b1-4)[Neu5Ac(a2-6)]Gal(b1-4)Glc',\n",
       " 'GalNAc(b1-6)GalNAc',\n",
       " 'GalNAc3S(b1-4)Gal3S(b1-4)Glc',\n",
       " 'GalNAc3S(b1-4)GlcNAc',\n",
       " 'GalNAc3S(b1-4)GlcNAc3S',\n",
       " 'GalNAc3S(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'GalNAc4S6S(b1-4)GlcNAc',\n",
       " 'GalNAc6S(b1-4)GlcNAc',\n",
       " 'Glc(a1-2)Glc',\n",
       " 'Glc(a1-2)Glc(a1-2)Glc',\n",
       " 'Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc',\n",
       " 'Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc',\n",
       " 'Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc',\n",
       " 'Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc',\n",
       " 'Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc',\n",
       " 'Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc(a1-2)Glc',\n",
       " 'Glc(a1-2)Glc(a1-3)Glc(a1-3)Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc',\n",
       " 'Glc(a1-2)Glc(a1-3)Glc(a1-3)Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)[Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc',\n",
       " 'Glc(a1-3)Glc',\n",
       " 'Glc(a1-3)Glc(a1-3)Glc',\n",
       " 'Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc',\n",
       " 'Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc',\n",
       " 'Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc',\n",
       " 'Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc',\n",
       " 'Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc',\n",
       " 'Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc',\n",
       " 'Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc',\n",
       " 'Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc',\n",
       " 'Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc',\n",
       " 'Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc(a1-3)Glc',\n",
       " 'Glc(a1-3)Glc(a1-3)Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Glc(a1-3)Glc(a1-3)Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)[Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc',\n",
       " 'Glc(a1-3)Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Glc(a1-4)Glc',\n",
       " 'Glc(a1-4)Glc(a1-4)Glc',\n",
       " 'Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc',\n",
       " 'Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc',\n",
       " 'Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc',\n",
       " 'Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc',\n",
       " 'Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc',\n",
       " 'Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc',\n",
       " 'Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc',\n",
       " 'Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc',\n",
       " 'Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc',\n",
       " 'Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc',\n",
       " 'Glc(a1-4)Glc(a1-6)Glc',\n",
       " 'Glc(a1-6)Glc',\n",
       " 'Glc(a1-6)Glc(a1-4)Glc',\n",
       " 'Glc(a1-6)Glc(a1-4)Glc(a1-4)Glc',\n",
       " 'Glc(a1-6)Glc(a1-4)Glc(a1-4)Glc(a1-6)Glc(a1-4)Glc(a1-4)Glc',\n",
       " 'Glc(a1-6)Glc(a1-6)Glc',\n",
       " 'Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc',\n",
       " 'Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc',\n",
       " 'Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc',\n",
       " 'Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc',\n",
       " 'Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc',\n",
       " 'Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc',\n",
       " 'Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc',\n",
       " 'Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc',\n",
       " 'Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc',\n",
       " 'Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc(a1-6)Glc',\n",
       " 'Glc(b1-2)Glc',\n",
       " 'Glc(b1-2)Glc(b1-2)Glc',\n",
       " 'Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc',\n",
       " 'Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc',\n",
       " 'Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc',\n",
       " 'Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc',\n",
       " 'Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc',\n",
       " 'Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc',\n",
       " 'Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc',\n",
       " 'Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc',\n",
       " 'Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc',\n",
       " 'Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc(b1-2)Glc',\n",
       " 'Glc(b1-3)Glc',\n",
       " 'Glc(b1-3)Glc(b1-3)Glc',\n",
       " 'Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc',\n",
       " 'Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc',\n",
       " 'Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc',\n",
       " 'Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc',\n",
       " 'Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc',\n",
       " 'Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc',\n",
       " 'Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc',\n",
       " 'Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc',\n",
       " 'Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc',\n",
       " 'Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc',\n",
       " 'Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)[Glc(b1-6)]Glc(b1-3)Glc',\n",
       " 'Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)[Glc(b1-6)]Glc(b1-3)Glc(b1-3)Glc',\n",
       " 'Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)[Glc(b1-6)]Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc',\n",
       " 'Glc(b1-3)Glc(b1-3)Glc(b1-3)[Glc(b1-6)]Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)[Glc(b1-6)]Glc(b1-3)Glc',\n",
       " 'Glc(b1-3)Glc(b1-3)[Glc(b1-6)]Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc',\n",
       " 'Glc(b1-3)Glc(b1-4)Glc',\n",
       " 'Glc(b1-3)Glc(b1-4)Glc(b1-4)Glc',\n",
       " 'Glc(b1-3)Glc(b1-4)Glc(b1-6)Glc(b1-4)Glc',\n",
       " 'Glc(b1-3)[Glc(b1-6)]Glc(a1-3)Glc(b1-3)[Glc(b1-6)]Glc',\n",
       " 'Glc(b1-3)[Glc(b1-6)]Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc(b1-3)Glc',\n",
       " 'Glc(b1-4)Glc',\n",
       " 'Glc(b1-4)Glc(b1-3)Glc',\n",
       " 'Glc(b1-4)Glc(b1-3)Glc(b1-4)Glc',\n",
       " 'Glc(b1-4)Glc(b1-4)Glc',\n",
       " 'Glc(b1-4)Glc(b1-4)Glc(b1-3)Glc',\n",
       " 'Glc(b1-4)Glc(b1-4)Glc(b1-3)Glc(b1-4)Glc(b1-4)Glc(b1-3)Glc',\n",
       " 'Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc',\n",
       " 'Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-3)Glc',\n",
       " 'Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc',\n",
       " 'Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc',\n",
       " 'Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc',\n",
       " 'Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc',\n",
       " 'Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc',\n",
       " 'Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc',\n",
       " 'Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc',\n",
       " 'Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc',\n",
       " 'Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc(b1-4)Glc',\n",
       " 'Glc(b1-6)Glc',\n",
       " 'Glc(b1-6)Glc(b1-6)Glc',\n",
       " 'Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc',\n",
       " 'Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc',\n",
       " 'Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc',\n",
       " 'Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc',\n",
       " 'Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc',\n",
       " 'Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc',\n",
       " 'Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc',\n",
       " 'Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc',\n",
       " 'Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc',\n",
       " 'Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc',\n",
       " 'Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc(b1-6)Glc',\n",
       " 'GlcA(b1-3)Gal',\n",
       " 'GlcA(b1-3)Gal(b1-3)GlcNAc',\n",
       " 'GlcA(b1-3)Gal(b1-4)Glc',\n",
       " 'GlcA(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'GlcA(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'GlcA(b1-3)GlcNAc',\n",
       " 'GlcA(b1-3)GlcNAc(b1-4)GlcA(b1-3)GlcNAc',\n",
       " 'GlcA(b1-3)GlcNAc(b1-4)GlcA(b1-3)GlcNAc(b1-4)GlcA(b1-3)GlcNAc(b1-4)GlcA(b1-3)GlcNAc(b1-4)GlcA(b1-3)GlcNAc(b1-4)GlcA(b1-3)GlcNAc(b1-4)GlcA(b1-3)GlcNAc',\n",
       " 'GlcA(b1-3)GlcNAc(b1-4)GlcA(b1-3)GlcNAc(b1-4)GlcA(b1-3)GlcNAc(b1-4)GlcA(b1-3)GlcNAc(b1-4)GlcA(b1-3)GlcNAc(b1-4)GlcA(b1-3)GlcNAc(b1-4)GlcA(b1-3)GlcNAc(b1-4)GlcA(b1-3)GlcNAc',\n",
       " 'GlcA(b1-4)GlcNAc(a1-4)GlcA(b1-4)2,5-Anhydro-Man',\n",
       " 'GlcA(b1-4)GlcNAc(a1-4)GlcA(b1-4)GlcNAc(a1-4)GlcA(b1-4)GlcNAc(a1-4)GlcA(b1-4)2,5-Anhydro-Man',\n",
       " 'GlcA(b1-6)Gal',\n",
       " 'GlcA3S(b1-3)Gal',\n",
       " 'GlcA3S(b1-3)Gal(b1-4)Glc',\n",
       " 'GlcA3S(b1-3)Gal(b1-4)GlcNAc(b1-2)Man',\n",
       " 'GlcA3S(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'GlcA3S(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'GlcNAc(a1-3)Gal(b1-4)GlcNAc',\n",
       " 'GlcNAc(a1-4)Gal',\n",
       " 'GlcNAc(a1-4)Gal(b1-3)GlcNAc',\n",
       " 'GlcNAc(a1-4)Gal(b1-3)[Fuc(a1-2)]Gal',\n",
       " 'GlcNAc(a1-4)Gal(b1-4)GalNAc',\n",
       " 'GlcNAc(a1-4)Gal(b1-4)GlcNAc',\n",
       " 'GlcNAc(a1-4)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'GlcNAc(a1-4)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'GlcNAc(a1-4)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'GlcNAc(a1-4)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'GlcNAc(a1-6)Gal(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-2)Fuc',\n",
       " 'GlcNAc(b1-2)Gal(b1-3)GalNAc',\n",
       " 'GlcNAc(b1-2)Gal(b1-3)GlcNAc',\n",
       " 'GlcNAc(b1-2)Man',\n",
       " 'GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)Man(a1-6)]Man',\n",
       " 'GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)Man(a1-6)][GlcNAc(b1-4)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-4)][GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-4)][GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'GlcNAc(b1-2)Man(a1-3)[Man(a1-3)[Man(a1-6)]Man(a1-6)][GlcNAc(b1-4)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-2)[GlcNAc(b1-4)]Man(a1-3)[GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-2)[GlcNAc(b1-4)]Man(a1-3)[GlcNAc(b1-2)Man(a1-6)][GlcNAc(b1-4)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-2)[GlcNAc(b1-4)]Man(a1-3)[GlcNAc(b1-2)[GlcNAc(b1-4)][GlcNAc(b1-6)]Man(a1-6)][GlcNAc(b1-4)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-2)[GlcNAc(b1-4)]Man(a1-3)[GlcNAc(b1-2)[GlcNAc(b1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-2)[GlcNAc(b1-4)]Man(a1-3)[GlcNAc(b1-4)][GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-2)[GlcNAc(b1-4)]Man(a1-3)[GlcNAc(b1-4)][GlcNAc(b1-2)[GlcNAc(b1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-2)[GlcNAc(b1-4)]Man(a1-3)[GlcNAc(b1-4)][GlcNAc(b1-2)[GlcNAc(b1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'GlcNAc(b1-2)[GlcNAc(b1-6)]Man(a1-6)[GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-2)[GlcNAc(b1-6)]Man(a1-6)[GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'GlcNAc(b1-2)[GlcNAc(b1-6)]Man(a1-6)[GlcNAc(b1-2)Man(a1-3)][GlcNAc(b1-4)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-2)[GlcNAc(b1-6)]Man(a1-6)[GlcNAc(b1-4)][GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-3)Fuc',\n",
       " 'GlcNAc(b1-3)Gal',\n",
       " 'GlcNAc(b1-3)Gal(b1-3)GalNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)[GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)[GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)[GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)[GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)GalNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)GalNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)[GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)]GalNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)[Gal(b1-3)]GalNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)[GlcNAc(b1-3)Gal(b1-3)]GalNAc',\n",
       " 'GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)[GlcNAc(b1-3)]Gal(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-3)GalNAc',\n",
       " 'GlcNAc(b1-3)Man',\n",
       " 'GlcNAc(b1-3)[GlcNAc(b1-4)][GlcNAc(b1-6)]GlcNAc',\n",
       " 'GlcNAc(b1-3)[GlcNAc(b1-6)]Gal',\n",
       " 'GlcNAc(b1-3)[GlcNAc(b1-6)]Gal(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-3)[GlcNAc(b1-6)]Gal(b1-4)GlcNAc(b1-3)[GlcNAc(b1-6)]Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'GlcNAc(b1-3)[GlcNAc(b1-6)]GalNAc',\n",
       " 'GlcNAc(b1-3)[GlcNAc(b1-6)]GlcNAc',\n",
       " 'GlcNAc(b1-3)[GlcNAc(b1-6)]GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-4)Fuc',\n",
       " 'GlcNAc(b1-4)Gal(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-4)[GlcNAc(b1-6)]GalNAc',\n",
       " 'GlcNAc(b1-6)Gal(b1-4)GlcNAc',\n",
       " 'GlcNAc(b1-6)GalNAc',\n",
       " 'GlcNAc3Prop(b1-4)GlcNAc',\n",
       " 'GlcNAc6S(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'HexA(a1-3)GalNAc4S(b1-4)GlcA(b1-3)GalNAc4S',\n",
       " 'HexA(a1-3)GalNAc4S(b1-4)GlcA(b1-3)GalNAc4S(b1-4)GlcA(b1-3)GalNAc4S(b1-4)GlcA(b1-3)GalNAc4S(b1-4)GlcA(b1-3)GalNAc4S(b1-4)GlcA(b1-3)GalNAc4S(b1-4)GlcA(b1-3)GalNAc4S',\n",
       " 'HexA(a1-3)GalNAc4S(b1-4)GlcA(b1-3)GalNAc4S(b1-4)GlcA(b1-3)GalNAc4S(b1-4)GlcA(b1-3)GalNAc4S(b1-4)GlcA(b1-3)GalNAc4S(b1-4)GlcA(b1-3)GalNAc4S(b1-4)GlcA(b1-3)GalNAc4S(b1-4)GlcA(b1-3)GalNAc4S',\n",
       " 'HexA(a1-3)GalNAc4S(b1-4)IdoA(a1-3)GalNAc4S',\n",
       " 'HexA(a1-3)GalNAc4S(b1-4)IdoA(a1-3)GalNAc4S(b1-4)IdoA(a1-3)GalNAc4S(b1-4)IdoA(a1-3)GalNAc4S(b1-4)IdoA(a1-3)GalNAc4S(b1-4)IdoA(a1-3)GalNAc4S(b1-4)IdoA(a1-3)GalNAc4S',\n",
       " 'HexA(a1-3)GalNAc4S(b1-4)IdoA(a1-3)GalNAc4S(b1-4)IdoA(a1-3)GalNAc4S(b1-4)IdoA(a1-3)GalNAc4S(b1-4)IdoA(a1-3)GalNAc4S(b1-4)IdoA(a1-3)GalNAc4S(b1-4)IdoA(a1-3)GalNAc4S(b1-4)IdoA(a1-3)GalNAc4S',\n",
       " 'HexA(a1-3)GalNAc6S(b1-4)GlcA(b1-3)GalNAc6S',\n",
       " 'HexA(a1-3)GalNAc6S(b1-4)GlcA(b1-3)GalNAc6S(b1-4)GlcA(b1-3)GalNAc6S(b1-4)GlcA(b1-3)GalNAc6S(b1-4)GlcA(b1-3)GalNAc6S(b1-4)GlcA(b1-3)GalNAc6S(b1-4)GlcA(b1-3)GalNAc6S',\n",
       " 'HexA(a1-3)GalNAc6S(b1-4)GlcA(b1-3)GalNAc6S(b1-4)GlcA(b1-3)GalNAc6S(b1-4)GlcA(b1-3)GalNAc6S(b1-4)GlcA(b1-3)GalNAc6S(b1-4)GlcA(b1-3)GalNAc6S(b1-4)GlcA(b1-3)GalNAc6S(b1-4)GlcA(b1-3)GalNAc6S',\n",
       " 'HexA(a1-4)GlcNAc(a1-4)HexA(a1-4)GlcNAc(a1-4)HexA(a1-4)GlcNAc(a1-4)HexA(a1-4)Man',\n",
       " 'HexA2S(a1-4)GlcNS6S',\n",
       " 'HexA2S(a1-4)GlcNS6S(a1-4)IdoA2S(a1-4)GlcNS6S(a1-4)IdoA2S(a1-4)GlcNS6S(a1-4)IdoA2S(a1-4)GlcNS6S(a1-4)IdoA2S(a1-4)GlcNS6S(a1-4)IdoA2S(a1-4)GlcNS6S(a1-4)IdoA2S(a1-4)GlcNS6S(a1-4)IdoA2S(a1-4)GlcNS6S(a1-4)IdoA2S',\n",
       " 'HexA6S(a1-4)GlcNS6S(a1-4)IdoA2S(a1-4)GlcNS6S',\n",
       " 'HexA6S(a1-4)GlcNS6S(a1-4)IdoA2S(a1-4)GlcNS6S(a1-4)IdoA2S(a1-4)GlcNS6S(a1-4)IdoA2S(a1-4)GlcNS6S(a1-4)IdoA2S(a1-4)GlcNS6S(a1-4)IdoA2S(a1-4)GlcNS6S(a1-4)IdoA2S(a1-4)GlcNS6S',\n",
       " 'Kdn(a2-3)Gal(b1-3)GalNAc',\n",
       " 'Kdn(a2-3)Gal(b1-3)GalNAc(b1-4)[Kdn(a2-3)]Gal(b1-4)Glc',\n",
       " 'Kdn(a2-3)Gal(b1-3)GlcNAc',\n",
       " 'Kdn(a2-3)Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Kdn(a2-3)Gal(b1-4)Glc',\n",
       " 'Kdn(a2-3)Gal(b1-4)GlcNAc',\n",
       " 'Kdn(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Kdn(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal',\n",
       " 'Kdn(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Kdn(a2-3)Gal(b1-4)[Rha(a1-3)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Kdn(a2-3)[GalNAc(b1-4)]Gal(b1-3)GlcNAc',\n",
       " 'Kdn(a2-3)[GalNAc(b1-4)]Gal(b1-4)Glc',\n",
       " 'Kdn(a2-6)Gal',\n",
       " 'Kdn(a2-6)Gal(b1-4)GlcNAc',\n",
       " 'Man(a1-2)Man',\n",
       " 'Man(a1-2)Man(a1-2)Man(a1-3)Man',\n",
       " 'Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)[Man(a1-3)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)[Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Man(a1-2)Man(a1-2)Man(a1-6)[Man(a1-3)]Man',\n",
       " 'Man(a1-2)Man(a1-3)Man',\n",
       " 'Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)[Man(a1-3)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)]Man',\n",
       " 'Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)]Man(a1-6)[Man(a1-2)Man(a1-2)Man(a1-3)]Man',\n",
       " 'Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)]Man(a1-6)[Man(a1-2)Man(a1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)]Man(a1-6)[Man(a1-3)[Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Man(a1-2)Man(a1-3)[Man(a1-3)[Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Man(a1-2)Man(a1-3)[Man(a1-6)]Man(a1-6)[Man(a1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Man(a1-2)Man(a1-6)[Man(a1-3)]Man(a1-6)[Man(a1-2)Man(a1-2)Man(a1-3)]Man',\n",
       " 'Man(a1-2)Man(a1-6)[Man(a1-3)]Man(a1-6)[Man(a1-2)Man(a1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Man(a1-3)Man',\n",
       " 'Man(a1-3)Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Man(a1-3)Man(b1-4)GlcNAc',\n",
       " 'Man(a1-3)[Man(a1-6)]Man',\n",
       " 'Man(a1-3)[Man(a1-6)]Man(a1-6)Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Man(a1-3)[Man(a1-6)]Man(a1-6)[Man(a1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Man(a1-3)[Man(a1-6)]Man(a1-6)[Man(a1-3)]Man',\n",
       " 'Man(a1-3)[Man(a1-6)]Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Man(a1-6)Man',\n",
       " 'Man(a1-6)Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Man(b1-4)GlcNAc',\n",
       " 'Man(b1-4)Man(b1-4)Man(b1-4)Man',\n",
       " 'Man(b1-4)Man(b1-4)Man(b1-4)Man(b1-4)Man',\n",
       " 'Man(b1-4)Man(b1-4)Man(b1-4)Man(b1-4)Man(b1-4)Man',\n",
       " 'Man6P(a1-3)Man(a1-3)Man(a1-3)Man(a1-2)Man',\n",
       " 'MurNAc(b1-4)GlcNAc',\n",
       " 'Neu(a2-3)Gal(b1-4)Glc',\n",
       " 'Neu(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu(a2-3)Gal(b1-4)GlcNAc6S(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc6S(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu(a2-3)Gal6S(b1-4)[Fuc(a1-3)]GlcN(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu(a2-6)Gal(b1-4)Glc',\n",
       " 'Neu4Ac5Ac(a2-3)Gal(b1-4)Glc',\n",
       " 'Neu4Me5Ac(a2-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-2)Glc',\n",
       " 'Neu5Ac(a2-3)4dGal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal',\n",
       " 'Neu5Ac(a2-3)6dGal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal',\n",
       " 'Neu5Ac(a2-3)Gal',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)6dGalNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GalNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GalNAc(b1-3)Gal(a1-4)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GalNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GalNAc(b1-4)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GalNAc(b1-4)[Neu5Ac(a2-3)]Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GalNAc(b1-4)[Neu5Ac(a2-8)Neu5Ac(a2-3)]Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GalNAc6S',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GalNAc6S(b1-4)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GlcNAc(b1-2)Man',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-4)][Neu5Ac(a2-3)Gal(b1-3)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GlcNAc(b1-2)Man(a1-3)[Neu5Ac(a2-3)Gal(b1-3)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GlcNAc(b1-2)[Neu5Ac(a2-3)Gal(b1-3)GlcNAc(b1-4)]Man(a1-6)[Neu5Ac(a2-3)Gal(b1-3)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GlcNAc(b1-3)Gal(b1-3)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GlcNAc(b1-3)GalNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GlcNAc(b1-3)[Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GlcNAc(b1-4)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GlcNAc(b1-6)GalNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)GlcNAc6S',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)[Fuc(a1-3)]GlcN(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)[Fuc(a1-4)]GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)Gal(b1-3)[Fuc(a1-4)]GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)[Fuc(a1-4)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)[GlcNAc(b1-6)]GalNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)[Neu5Ac(a2-3)Gal(b1-4)]GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)[Neu5Ac(a2-6)]GalNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)[Neu5Ac(a2-6)]GalNAc(b1-4)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)[Neu5Ac(a2-6)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)[Neu5Ac(a2-6)]GlcNAc(b1-3)[Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-3)]Neu5Ac(a2-3)Gal(b1-4)]GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)Man(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-4)][Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)]Man(a1-6)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)][GlcNAc(b1-4)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)[GlcNAc(b1-4)][Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)[GlcNAc(b1-4)][Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-4)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)][GlcNAc(b1-4)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-3)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]GalNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)GalNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)]GalNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)[Gal(b1-3)]GalNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)GalNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-3)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-6)]GalNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-6)GalNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-6)GalNAc(a1-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-6)[Gal(b1-3)]GalNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-6)[Neu5Ac(a2-3)Gal(b1-3)]GalNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc6S',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc6S(b1-3)Gal6S(b1-3)GlcNAc6S',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[3dFuc(a1-3)]GlcNAc(b1-3)Gal',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[4dFuc(a1-3)]GlcNAc(b1-3)Gal',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[6dTal(a1-3)]GlcNAc(b1-3)Gal',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)GalNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-4)GalNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-6)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-6)[Gal(b1-3)]GalNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-6)[Neu5Ac(a2-3)Gal(b1-3)]GalNAc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc6S',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc6S(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc6S(b1-3)Gal6S(b1-3)GlcNAc6S',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Fuc2Me(a1-3)]GlcNAc(b1-3)Gal',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Qui(a1-3)]GlcNAc(b1-3)Gal',\n",
       " 'Neu5Ac(a2-3)Gal(b1-4)[Rha(a1-3)]GlcNAc(b1-3)Gal',\n",
       " 'Neu5Ac(a2-3)Gal3S',\n",
       " 'Neu5Ac(a2-3)Gal6S(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal6S(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Neu5Ac(a2-3)Gal6S(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)Gal6S(b1-4)[Fuc(a1-3)]GlcNAc6S(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)GalNAc',\n",
       " 'Neu5Ac(a2-3)GalNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)GalNBz(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)[GalNAc(b1-4)]Gal(b1-3)Gal',\n",
       " 'Neu5Ac(a2-3)[GalNAc(b1-4)]Gal(b1-3)GalNAc(b1-4)[Neu5Ac(a2-3)]Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)[GalNAc(b1-4)]Gal(b1-3)Glc',\n",
       " 'Neu5Ac(a2-3)[GalNAc(b1-4)]Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)[GalNAc(b1-4)]Gal(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-3)[GalNAc(b1-4)]Gal(b1-4)GlcNAc(b1-3)GalNAc',\n",
       " 'Neu5Ac(a2-3)[GalNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-3)[Neu5Ac(a2-6)]GalNAc',\n",
       " 'Neu5Ac(a2-6)Gal',\n",
       " 'Neu5Ac(a2-6)Gal(b1-3)GalNAc(b1-4)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-3)GlcNAc(b1-3)[Fuc(a1-3)[Gal(b1-4)]GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-3)GlcNAc(b1-3)[Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-3)[Neu5Ac(a2-6)]GalNAc(b1-4)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-2)Man(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[GlcNAc(b1-4)][Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Gc(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)[Gal(b1-4)GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)[GlcNAc(b1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)[Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)[GlcNAc(b1-4)][Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)[GlcNAc(b1-4)][Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-4)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)][GlcNAc(b1-4)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-3)Gal(b1-3)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]GalNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]GalNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)GalNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-3)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)]GalNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-6)[Gal(b1-3)]GalNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]Glc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-3)GalNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-3)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-6)]Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-3)[Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-6)]GalNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-6)GalNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-6)[Gal(b1-3)GlcNAc(b1-3)]Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc(b1-6)[Gal(b1-3)]GalNAc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)GlcNAc6S',\n",
       " 'Neu5Ac(a2-6)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-6)Gal(b1-6)GalNAc(b1-4)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-6)Gal6S',\n",
       " 'Neu5Ac(a2-6)Gal6S(b1-4)Glc',\n",
       " 'Neu5Ac(a2-6)Gal6S(b1-4)Glc4S',\n",
       " 'Neu5Ac(a2-6)GalNAc',\n",
       " 'Neu5Ac(a2-6)GalNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)GalNAc(b1-4)GlcNAc6S',\n",
       " 'Neu5Ac(a2-6)Glc',\n",
       " 'Neu5Ac(a2-6)Glc6S',\n",
       " 'Neu5Ac(a2-6)GlcNAc',\n",
       " 'Neu5Ac(a2-6)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)GlcNAc(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-6)GlcNAc6S',\n",
       " 'Neu5Ac(a2-8)Neu5Ac',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-3)Gal',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-3)Gal(b1-3)GalNAc(b1-4)[Neu5Ac(a2-3)]Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-3)Gal(b1-3)GalNAc(b1-4)[Neu5Ac(a2-8)Neu5Ac(a2-3)]Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-3)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-3)[Gal(b1-3)GalNAc(b1-4)]Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-3)[GalNAc(b1-4)]Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-6)Gal',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-6)Glc',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-3)[Gal(b1-3)GalNAc(b1-4)]Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-3)[GalNAc(b1-4)]Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-6)Gal',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-3)[GalNAc(b1-4)]Gal(b1-4)Glc',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac',\n",
       " 'Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac(a2-8)Neu5Ac',\n",
       " 'Neu5Ac(a2-8)Neu5Gc(a2-3)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Ac(a2-9)Neu5Ac(a2-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(b1-6)GalNAc',\n",
       " 'Neu5Ac(b2-3)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(b2-6)Gal(b1-4)Glc',\n",
       " 'Neu5Ac(b2-6)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Ac(b2-6)GalNAc',\n",
       " 'Neu5Ac9Ac(a2-3)Gal(b1-3)GlcNAc',\n",
       " 'Neu5Ac9Ac(a2-3)Gal(b1-3)[Fuc(a1-4)]GlcNAc',\n",
       " 'Neu5Ac9Ac(a2-3)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Ac9Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Neu5Ac9Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal',\n",
       " 'Neu5Ac9Ac(a2-6)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Ac9Me(a2-3)Gal(b1-4)Glc',\n",
       " 'Neu5Gc(a2-3)Gal(b1-3)GlcNAc',\n",
       " 'Neu5Gc(a2-3)Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu5Gc(a2-3)Gal(b1-3)[Fuc(a1-4)]GlcNAc',\n",
       " 'Neu5Gc(a2-3)Gal(b1-4)Glc',\n",
       " 'Neu5Gc(a2-3)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Gc(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Gc(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Gc(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Neu5Gc(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-3)Gal',\n",
       " 'Neu5Gc(a2-3)Gal(b1-4)[Fuc(a1-4)]GlcNAc',\n",
       " 'Neu5Gc(a2-3)[GalNAc(b1-4)]Gal',\n",
       " 'Neu5Gc(a2-6)Gal(b1-3)GlcNAc(b1-3)Gal(b1-4)Glc',\n",
       " 'Neu5Gc(a2-6)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Gc(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Gc(a2-6)Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Neu5Gc(a2-6)GalNAc',\n",
       " 'Neu5Gc(a2-8)Neu5Ac(a2-3)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Gc(a2-8)Neu5Gc(a2-3)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Gc(a2-8)Neu5Gc(a2-3)Gal(b1-4)GlcNAc(b1-3)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Gc(a2-8)Neu5Gc(a2-6)Gal(b1-4)GlcNAc',\n",
       " 'Neu5Gc(b2-6)Gal(b1-4)GlcNAc',\n",
       " 'Rha(a1-6)Glc',\n",
       " 'Xyl(a1-6)Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)[Xyl(a1-6)]Glc(b1-4)Glc',\n",
       " 'Xyl(b1-2)[Man(a1-3)][Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',\n",
       " 'Xyl(b1-2)[Man(a1-3)][Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-3)]GlcNAc',\n",
       " 'Xyl(b1-4)Xyl(b1-4)Xyl(b1-4)Xyl(b1-4)Xyl',\n",
       " 'Xyl(b1-4)Xyl(b1-4)Xyl(b1-4)Xyl(b1-4)Xyl(b1-4)Xyl']"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11381855-f688-4804-bb7b-e51ca1fa97c1",
   "metadata": {
    "id": "11381855-f688-4804-bb7b-e51ca1fa97c1"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "gpuType": "T4",
   "machine_shape": "hm",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
